diff --git a/htdocs/api/class/api.class.php b/htdocs/api/class/api.class.php
index 657a97e2e17..7e6c19a1b7b 100644
--- a/htdocs/api/class/api.class.php
+++ b/htdocs/api/class/api.class.php
@@ -63,17 +63,6 @@ class DolibarrApi
 
 		$production_mode = getDolGlobalBool('API_PRODUCTION_MODE');
 
-		if ($production_mode) {
-			// Create the directory Defaults::$cacheDirectory if it does not exist. If dir does not exist, using production_mode generates an error 500.
-			include_once DOL_DOCUMENT_ROOT.'/core/lib/files.lib.php';
-			if (!dol_is_dir(Defaults::$cacheDirectory)) {
-				dol_mkdir(Defaults::$cacheDirectory, DOL_DATA_ROOT);
-			}
-			if (getDolGlobalString('MAIN_API_DEBUG')) {
-				dol_syslog("Debug API construct::cacheDirectory=".Defaults::$cacheDirectory, LOG_DEBUG, 0, '_api');
-			}
-		}
-
 		$this->r = new Restler($production_mode, $refreshCache);
 
 		$urlwithouturlroot = preg_replace('/'.preg_quote(DOL_URL_ROOT, '/').'$/i', '', trim($dolibarr_main_url_root));
diff --git a/htdocs/api/class/api_access.class.php b/htdocs/api/class/api_access.class.php
index 974bd5703be..38d44b2bfae 100644
--- a/htdocs/api/class/api_access.class.php
+++ b/htdocs/api/class/api_access.class.php
@@ -272,14 +272,6 @@ class DolibarrApiAccess implements iAuthenticate
 				throw new RestException(401, $genericmessageerroruser);
 			}
 
-			// Check if user status is enabled
-			if ($fuser->status != $fuser::STATUS_ENABLED) {
-				// Status is disabled
-				dol_syslog("functions_isallowed::check_user_api_key Authentication KO for '".$login."': The user has been disabled", LOG_NOTICE);
-				sleep(1); // Anti brute force protection. Must be same delay when user and password are not valid.
-				throw new RestException(401, $genericmessageerroruser);
-			}
-
 			// Check if session was unvalidated by a password change
 			if (($fuser->flagdelsessionsbefore && !empty($_SESSION["dol_logindate"]) && $fuser->flagdelsessionsbefore > $_SESSION["dol_logindate"])) {
 				// Session is no more valid
diff --git a/htdocs/api/class/api_documents.class.php b/htdocs/api/class/api_documents.class.php
index fe5c51343bf..ea59fc7c253 100644
--- a/htdocs/api/class/api_documents.class.php
+++ b/htdocs/api/class/api_documents.class.php
@@ -607,11 +607,6 @@ class Documents extends DolibarrApi
 		$recursive = 0;
 		$type = 'files';
 
-		$objectType = $modulepart;
-		if (! empty($object->id) && ! empty($object->table_element)) {
-			$objectType = $object->table_element;
-		}
-
 		$filearray = dol_dir_list($upload_dir, $type, $recursive, '', '(\.meta|_preview.*\.png)$', $sortfield, (strtolower($sortorder) == 'desc' ? SORT_DESC : SORT_ASC), 1);
 
 		$countarray = is_array($filearray) ? count($filearray) : 0;
@@ -623,53 +618,11 @@ class Documents extends DolibarrApi
 			if (($object->id) > 0 && !empty($modulepart)) {
 				require_once DOL_DOCUMENT_ROOT.'/ecm/class/ecmfiles.class.php';
 				$ecmfile = new EcmFiles($this->db);
-				$result = $ecmfile->fetchAll('', '', 0, 0, array('t.src_object_type' => $objectType, 't.src_object_id' => $object->id));
+				$result = $ecmfile->fetchAll('', '', 0, 0, array('t.src_object_type' => $modulepart, 't.src_object_id' => $object->id));
 				if ($result < 0) {
 					throw new RestException(503, 'Error when retrieve ecm list : '.$this->db->lasterror());
 				} elseif (is_array($ecmfile->lines) && count($ecmfile->lines) > 0) {
-					foreach ($filearray as &$fileitem) {
-						foreach ($ecmfile->lines as $line) {
-							if ($fileitem['name'] == $line->filename) {
-								// Next line converts EcmFilesLine properties to array
-								//$fileitem = array_merge($fileitem, (array) $line);
-								$fileitem['ref'] = $line->ref;
-								$fileitem['label'] = $line->label;
-								$fileitem['filepath'] = $line->filepath;
-								$fileitem['filename'] = $line->filename;
-								$fileitem['fullpath_orig'] = $line->fullpath_orig;
-								$fileitem['position'] = $line->position;
-								$fileitem['gen_or_uploaded'] = $line->gen_or_uploaded;
-								$fileitem['description'] = $line->desc;
-								$fileitem['keywords'] = $line->keywords;
-								$fileitem['cover'] = $line->cover;
-								$fileitem['share'] = $line->share;
-								$fileitem['date_c'] = $line->date_c;
-								$fileitem['agenda_id'] = $line->agenda_id;
-								$fileitem['fk_user_c'] = $line->fk_user_c;
-								$fileitem['fk_user_m'] = $line->fk_user_m;
-								$fileitem['note_private'] = $line->note_private;
-								$fileitem['note_public'] = $line->note_public;
-							}
-						}
-						if (isset($fileitem['relativename'])) {
-							$fileitem['content-type'] = dol_mimetype((string) $fileitem['relativename']);
-						}
-					}
-
-					// Select only files that match the requested $content_type, if provided
-					$arraycontenttype = explode(",", $content_type);
-					if (!empty($content_type)) {
-						$filearray = array_filter(
-							$filearray,
-							/**
-							 * @param array{name:string,path:string,level1name:string,relativename:string,fullname:string,date:string,size:int,perm:int,type:string,content-type?:string} $fileitem  Item to check for content type (for filtering)
-							 * @return bool True if file matches a content type
-							 */
-							static function ($fileitem) use (&$arraycontenttype) {
-								return in_array(($fileitem['content-type'] ?: 'UNKNOWN'), $arraycontenttype);
-							}
-						);
-					}
+					$filearray['ecmfiles_infos'] = $ecmfile->lines;
 				}
 			}
 		}
@@ -993,77 +946,16 @@ class Documents extends DolibarrApi
 			throw new RestException(500, "Failed to open file '".$destfiletmp."' for write");
 		}
 
-		$disablevirusscan = 0;
-		$src_file = $destfiletmp;
-		$dest_file = $destfile;
-
-		// Security:
-		// If we need to make a virus scan
-		if (empty($disablevirusscan) && file_exists($src_file)) {
-			$checkvirusarray = dolCheckVirus($src_file, $dest_file);
-			if (count($checkvirusarray)) {
-				dol_syslog('Files.lib::dol_move_uploaded_file File "'.$src_file.'" (target name "'.$dest_file.'") KO with antivirus: errors='.implode(',', $checkvirusarray), LOG_WARNING);
-				throw new RestException(500, 'ErrorFileIsInfectedWithAVirus: '.implode(',', $checkvirusarray));
-			}
-		}
-
-		// Security:
-		// Disallow file with some extensions. We rename them.
-		// Because if we put the documents directory into a directory inside web root (very bad), this allows to execute on demand arbitrary code.
-		if (isAFileWithExecutableContent($dest_file) && !getDolGlobalString('MAIN_DOCUMENT_IS_OUTSIDE_WEBROOT_SO_NOEXE_NOT_REQUIRED')) {
-			// $upload_dir ends with a slash, so be must be sure the medias dir to compare to ends with slash too.
-			$publicmediasdirwithslash = $conf->medias->multidir_output[$conf->entity];
-			if (!preg_match('/\/$/', $publicmediasdirwithslash)) {
-				$publicmediasdirwithslash .= '/';
-			}
-
-			if (strpos($upload_dir, $publicmediasdirwithslash) !== 0 || !getDolGlobalInt("MAIN_DOCUMENT_DISABLE_NOEXE_IN_MEDIAS_DIR")) {	// We never add .noexe on files into media directory
-				$dest_file .= '.noexe';
-			}
-		}
-
-		// Security:
-		// We refuse cache files/dirs, upload using .. and pipes into filenames.
-		if (preg_match('/^\./', basename($src_file)) || preg_match('/\.\./', $src_file) || preg_match('/[<>|]/', $src_file)) {
-			dol_syslog("Refused to deliver file ".$src_file, LOG_WARNING);
-			throw new RestException(500, "Refused to deliver file ".$src_file);
-		}
-
-		// Security:
-		// We refuse cache files/dirs, upload using .. and pipes into filenames.
-		if (preg_match('/^\./', basename($dest_file)) || preg_match('/\.\./', $dest_file) || preg_match('/[<>|]/', $dest_file)) {
-			dol_syslog("Refused to deliver file ".$dest_file, LOG_WARNING);
-			throw new RestException(500, "Refused to deliver file ".$dest_file);
-		}
-
-		$moreinfo = array('note_private' => 'File uploaded using API /documents from IP '.getUserRemoteIP());
-		// $object may be null
-		if (is_object($object) && $object->id > 0) {
-			$moreinfo['src_object_type'] = $object->table_element;
-			$moreinfo['src_object_id'] = $object->id;
-		}
-		if (!empty($array_options)) {
-			$moreinfo = array_merge($moreinfo, ["array_options" => $array_options]);
-		}
-		if (!empty($position)) {
-			$moreinfo = array_merge($moreinfo, ["position" => $position]);
-		}
-		if (!empty($cover)) {
-			$moreinfo = array_merge($moreinfo, ["cover" => $cover]);
-		}
-		$moreinfo['gen_or_uploaded'] = 'api';
-
-		// Move the temporary file at its final emplacement
-		$result = dol_move($destfiletmp, $dest_file, '0', $overwriteifexists, 1, 1, $moreinfo);
+		$result = dol_move($destfiletmp, $destfile, '0', $overwriteifexists, 0);
 		if (!$result) {
-			throw new RestException(500, "Failed to move file into '".$dest_file."'");
+			throw new RestException(500, "Failed to move file into '".$destfile."'");
 		}
 
 		if (is_object($object) && $generateThumbs) {
 			require_once DOL_DOCUMENT_ROOT.'/core/lib/files.lib.php';
 			require_once DOL_DOCUMENT_ROOT.'/core/lib/images.lib.php';	// image_format_supported() is defined here
-			if (image_format_supported($dest_file)) {
-				$object->addThumbs($dest_file);
+			if (image_format_supported($destfile)) {
+				$object->addThumbs($destfile);
 			}
 		}
 
diff --git a/htdocs/api/index.php b/htdocs/api/index.php
index e2d1c6f002d..f4262d51b44 100644
--- a/htdocs/api/index.php
+++ b/htdocs/api/index.php
@@ -68,27 +68,12 @@ if (!empty($_SERVER['HTTP_DOLAPIENTITY'])) {
 	define("DOLENTITY", (int) $_SERVER['HTTP_DOLAPIENTITY']);
 }
 
-// Response for preflight requests (used by browser when into a CORS context)
-if (!empty($_SERVER['REQUEST_METHOD']) && $_SERVER['REQUEST_METHOD'] == 'OPTIONS' && !empty($_SERVER['HTTP_ACCESS_CONTROL_REQUEST_HEADERS'])) {
-	header('Access-Control-Allow-Origin: *');
-	header('Access-Control-Allow-Methods: GET, POST, PUT, DELETE');
-	header('Access-Control-Allow-Headers: Content-Type, Authorization, api_key, DOLAPIKEY, DOLAPIENTITY');
-	http_response_code(204);
-	exit;
-}
-
 // When we request url to get the json file, we accept Cross site so we can include the descriptor into an external tool.
 if (preg_match('/\/explorer\/swagger\.json/', $_SERVER["PHP_SELF"])) {
 	header('Access-Control-Allow-Origin: *');
 	header('Access-Control-Allow-Methods: GET, POST, PUT, DELETE');
 	header('Access-Control-Allow-Headers: Content-Type, Authorization, api_key, DOLAPIKEY, DOLAPIENTITY');
 }
-// When we request url to get an API, we accept Cross site so we can make js API call inside another website
-if (preg_match('/\/api\/index\.php/', $_SERVER["PHP_SELF"])) {
-	header('Access-Control-Allow-Origin: *');
-	header('Access-Control-Allow-Methods: GET, POST, PUT, DELETE');
-	header('Access-Control-Allow-Headers: Content-Type, Authorization, api_key, DOLAPIKEY, DOLAPIENTITY');
-}
 header('X-Frame-Options: SAMEORIGIN');
 
 
@@ -381,7 +366,7 @@ if (!empty($reg[1]) && ($reg[1] != 'explorer' || ($reg[2] != '/swagger.json' &&
 
 		foreach ($listofendpoints as $endpointrule) {
 			$tmparray = explode(':', $endpointrule);
-			if (($classfile == $tmparray[0] || $classfile.'api' == $tmparray[0]) && $tmparray[1] == 1) {
+			if ($classfile == $tmparray[0] && $tmparray[1] == 1) {
 				$endpointisallowed = true;
 				break;
 			}
