diff --git a/htdocs/expedition/card.php b/htdocs/expedition/card.php
index 6dbcc538397..567a5f4a606 100644
--- a/htdocs/expedition/card.php
+++ b/htdocs/expedition/card.php
@@ -37,6 +37,7 @@
 require '../main.inc.php';
 require_once DOL_DOCUMENT_ROOT.'/core/class/html.formfile.class.php';
 require_once DOL_DOCUMENT_ROOT.'/expedition/class/expedition.class.php';
+require_once DOL_DOCUMENT_ROOT.'/expedition/class/ShipsGo_API.class.php';
 require_once DOL_DOCUMENT_ROOT.'/product/class/html.formproduct.class.php';
 require_once DOL_DOCUMENT_ROOT.'/core/lib/product.lib.php';
 require_once DOL_DOCUMENT_ROOT.'/core/lib/sendings.lib.php';
@@ -317,7 +318,7 @@ if (empty($reshook)) {
 
 		//var_dump($batch_line[2]);
 
-		if ($totalqty > 0 || !empty($conf->global->SHIPMENT_GETS_ALL_ORDER_PRODUCTS)) {		// There is at least one thing to ship
+		if ($totalqty > 0) {		// There is at least one thing to ship
 			//var_dump($_POST);exit;
 			for ($i = 0; $i < $num; $i++) {
 				$qty = "qtyl".$i;
@@ -327,7 +328,7 @@ if (empty($reshook)) {
 						//shipment from multiple stock locations
 						$nbstockline = count($stockLine[$i]);
 						for ($j = 0; $j < $nbstockline; $j++) {
-							if ($stockLine[$i][$j]['qty'] > 0 || ($stockLine[$i][$j]['qty'] == 0 && !empty($conf->global->SHIPMENT_GETS_ALL_ORDER_PRODUCTS))) {
+							if ($stockLine[$i][$j]['qty'] > 0) {
 								$ret = $object->addline($stockLine[$i][$j]['warehouse_id'], $stockLine[$i][$j]['ix_l'], $stockLine[$i][$j]['qty'], $array_options[$i]);
 								if ($ret < 0) {
 									setEventMessages($object->error, $object->errors, 'errors');
@@ -336,7 +337,7 @@ if (empty($reshook)) {
 							}
 						}
 					} else {
-						if (GETPOST($qty, 'int') > 0 || (GETPOST($qty, 'int') == 0 && !empty($conf->global->SHIPMENT_GETS_ALL_ORDER_PRODUCTS))) {
+						if (GETPOST($qty, 'int') > 0 || (GETPOST($qty, 'int') == 0 && $conf->global->SHIPMENT_GETS_ALL_ORDER_PRODUCTS)) {
 							$ent = "entl".$i;
 							$idl = "idl".$i;
 							$entrepot_id = is_numeric(GETPOST($ent, 'int')) ?GETPOST($ent, 'int') : GETPOST('entrepot_id', 'int');
@@ -410,9 +411,32 @@ if (empty($reshook)) {
 		|| (!empty($conf->global->MAIN_USE_ADVANCED_PERMS) && !empty($user->rights->expedition->shipping_advance->validate)))
 	) {
 		$object->fetch_thirdparty();
-
 		$result = $object->valid($user);
 
+		//ShipsGo Posting Container Info
+		$shipsGo = new ShipsGo_API($conf->global->API_KEY_SHIPSGO);
+		$ContainerNumber = $object->tracking_number;
+		$blno = $object->array_options['options_blno'];
+		$object->fetchObjectLinked();
+		$so_ref = end($object->linkedObjects['commande'])->ref;
+		$object->fetch_delivery_methods();		
+		$ShippingLine = $object->meths[$object->shipping_method_id];
+		$email = !empty($user->email) ? $user->email : '';
+		$Referance = $so_ref.' / '.$object->ref;
+		if (!empty($ContainerNumber) && !empty($ShippingLine) && !empty($Referance) && empty($object->array_options['options_requestid'])) {
+			if (!empty($blno)) {
+				$ship_result = $shipsGo->PostContainerInfoWithBl($ContainerNumber, $blno, $ShippingLine, $email, $Referance);
+			} else {
+				$ship_result = $shipsGo->PostContainerInfo($ContainerNumber, $ShippingLine, $email, $Referance);
+			}
+
+			$updatesql = "UPDATE ".MAIN_DB_PREFIX."expedition_extrafields SET";
+			$updatesql .= " requestid = ".$ship_result['RequestId'];
+			$updatesql .= ", updatedtime = '".date('Y-m-d H:i:s', dol_now())."'";
+			$updatesql .= " WHERE fk_object = ".$object->id;
+			$db->query($updatesql);
+		}
+		
 		if ($result < 0) {
 			setEventMessages($object->error, $object->errors, 'errors');
 		} else {
@@ -465,6 +489,11 @@ if (empty($reshook)) {
 		//		setEventMessages($object->error, $object->errors, 'errors');
 		//	}
 		//}
+	} elseif ($action == 'confirm_modif' && $confirm == 'yes' && $user->rights->expedition->creer &&$user->rights->expedition->shipping_advance->validate) { //SLY 2021.8.24
+		$result = $object->setStatut(0);
+		if ($result < 0) {
+			setEventMessages($object->error, $object->errors, 'errors');
+		}
 	} elseif ($action == 'setdate_livraison' && $user->rights->expedition->creer) {
 		//print "x ".$_POST['liv_month'].", ".$_POST['liv_day'].", ".$_POST['liv_year'];
 		$datedelivery = dol_mktime(GETPOST('liv_hour', 'int'), GETPOST('liv_min', 'int'), 0, GETPOST('liv_month', 'int'), GETPOST('liv_day', 'int'), GETPOST('liv_year', 'int'));
@@ -500,8 +529,8 @@ if (empty($reshook)) {
 			$object->trueWidth = trim(GETPOST('trueWidth', 'int'));
 		}
 		if ($action == 'settrueHeight') {
-						$object->trueHeight = trim(GETPOST('trueHeight', 'int'));
-						$object->size_units = GETPOST('size_units', 'int');
+			$object->trueHeight = trim(GETPOST('trueHeight', 'int'));
+			$object->size_units = GETPOST('size_units', 'int');
 		}
 		if ($action == 'settrueDepth') {
 			$object->trueDepth = trim(GETPOST('trueDepth', 'int'));
@@ -540,7 +569,6 @@ if (empty($reshook)) {
 		$object->fetch($id);
 		$lines = $object->lines;
 		$line = new ExpeditionLigne($db);
-		$line->fk_expedition = $object->id;
 
 		$num_prod = count($lines);
 		for ($i = 0; $i < $num_prod; $i++) {
@@ -581,9 +609,7 @@ if (empty($reshook)) {
 		$num_prod = count($lines);
 		for ($i = 0; $i < $num_prod; $i++) {
 			if ($lines[$i]->id == $line_id) {		// we have found line to update
-				$update_done = false;
 				$line = new ExpeditionLigne($db);
-				$line->fk_expedition = $object->id;
 
 				// Extrafields Lines
 				$line->array_options = $extrafields->getOptionalsFromPost($object->table_element_line);
@@ -622,8 +648,6 @@ if (empty($reshook)) {
 								if ($line->update($user) < 0) {
 									setEventMessages($line->error, $line->errors, 'errors');
 									$error++;
-								} else {
-									$update_done=true;
 								}
 							} else {
 								setEventMessages($lotStock->error, $lotStock->errors, 'errors');
@@ -666,8 +690,6 @@ if (empty($reshook)) {
 									if ($line->update($user) < 0) {
 										setEventMessages($line->error, $line->errors, 'errors');
 										$error++;
-									} else {
-										$update_done=true;
 									}
 								} else {
 									setEventMessages($line->error, $line->errors, 'errors');
@@ -685,8 +707,6 @@ if (empty($reshook)) {
 								if ($object->create_line_batch($line, $line->array_options) < 0) {
 									setEventMessages($object->error, $object->errors, 'errors');
 									$error++;
-								} else {
-									$update_done=true;
 								}
 							}
 						} else {
@@ -724,8 +744,6 @@ if (empty($reshook)) {
 										if ($line->update($user) < 0) {
 											setEventMessages($line->error, $line->errors, 'errors');
 											$error++;
-										} else {
-											$update_done=true;
 										}
 									}
 									unset($_POST[$stockLocation]);
@@ -740,8 +758,6 @@ if (empty($reshook)) {
 							if ($line->update($user) < 0) {
 								setEventMessages($line->error, $line->errors, 'errors');
 								$error++;
-							} else {
-								$update_done=true;
 							}
 							unset($_POST[$qty]);
 						}
@@ -754,17 +770,10 @@ if (empty($reshook)) {
 						if ($line->update($user) < 0) {
 							setEventMessages($line->error, $line->errors, 'errors');
 							$error++;
-						} else {
-							$update_done=true;
 						}
 						unset($_POST[$qty]);
 					}
 				}
-
-				if (empty($update_done)) {
-					$line->id = $lines[$i]->id;
-					$line->insertExtraFields();
-				}
 			}
 		}
 
@@ -1173,10 +1182,17 @@ if ($action == 'create') {
 						print "</td>\n";
 					}
 
+					// unit of order
+					$unit_order = '';
+					if ($conf->global->PRODUCT_USE_UNITS) {
+						$unit_order = measuringUnitString($line->fk_unit);
+					}
+
 					// Qty
+					$unit_order = measuringUnitString($line->fk_unit);
 					print '<td class="center">'.$line->qty;
 					print '<input name="qtyasked'.$indiceAsked.'" id="qtyasked'.$indiceAsked.'" type="hidden" value="'.$line->qty.'">';
-					print '</td>';
+					print ' '.$unit_order.'</td>';
 					$qtyProdCom = $line->qty;
 
 					// Qty already shipped
@@ -1184,7 +1200,7 @@ if ($action == 'create') {
 					$quantityDelivered = $object->expeditions[$line->id];
 					print $quantityDelivered;
 					print '<input name="qtydelivered'.$indiceAsked.'" id="qtydelivered'.$indiceAsked.'" type="hidden" value="'.$quantityDelivered.'">';
-					print '</td>';
+					print ' '.$unit_order.'</td>';
 
 					// Qty to ship
 					$quantityAsked = $line->qty;
@@ -1214,7 +1230,6 @@ if ($action == 'create') {
 								print '<input name="qtyl'.$indiceAsked.'" id="qtyl'.$indiceAsked.'" class="qtyl center" type="text" size="4" value="'.$deliverableQty.'">';
 							} else {
 								if (! empty($conf->global->SHIPMENT_GETS_ALL_ORDER_PRODUCTS)) {
-									print '<input name="idl'.$indiceAsked.'" type="hidden" value="'.$line->id.'">';
 									print '<input name="qtyl'.$indiceAsked.'" id="qtyl'.$indiceAsked.'" type="hidden" value="0">';
 								}
 
@@ -1465,10 +1480,6 @@ if ($action == 'create') {
 							}
 
 							foreach ($product->stock_warehouse as $warehouse_id => $stock_warehouse) {
-								if (!empty($warehousePicking) && !in_array($warehouse_id, $warehousePicking)) {
-									// if a warehouse was selected by user, picking is limited to this warehouse and his children.
-									continue;
-								}
 								$tmpwarehouseObject->fetch($warehouse_id);
 								if (($stock_warehouse->real > 0) && (count($stock_warehouse->detail_batch))) {
 									foreach ($stock_warehouse->detail_batch as $dbatch) {
@@ -1684,6 +1695,64 @@ if ($action == 'create') {
 		if ($action == 'cancel') {
 			$formconfirm = $form->formconfirm($_SERVER['PHP_SELF'].'?id='.$object->id, $langs->trans('CancelSending'), $langs->trans("ConfirmCancelSending", $object->ref), 'confirm_cancel', '', 0, 1);
 		}
+		// Confirm modification (back to draft status).
+		if ($action == 'modif') {
+			$formconfirm = $form->formconfirm($_SERVER['PHP_SELF'].'?id='.$object->id, $langs->trans('UnvalidateShipment'), $langs->trans("ConfirmUnvalidateShipment", $object->ref), 'confirm_modif', '', 0, 1);
+		}
+		// Update ShipsGo Status
+		if ($action == 'updateships') {
+			$ContainerNumber = $object->tracking_number;
+			$blno = $object->array_options['options_blno'];
+			$shipsGo = new ShipsGo_API($conf->global->API_KEY_SHIPSGO);
+			if (empty($object->array_options['options_requestid']) && (date($object->array_options['options_eta']) - dol_now()) > 864000) {
+				$object->fetchObjectLinked();
+				$so_ref = end($object->linkedObjects['commande'])->ref;
+				$object->fetch_delivery_methods();
+				$ShippingLine = $object->meths[$object->shipping_method_id];
+				$email = !empty($user->email) ? $user->email : '';
+				$Referance = $so_ref.' / '.$object->ref;
+				if (!empty($ContainerNumber) && !empty($ShippingLine) && !empty($Referance)) {
+					if (!empty($blno)) {
+						$ship_result = $shipsGo->PostContainerInfoWithBl($ContainerNumber, $blno, $ShippingLine, $email, $Referance);
+					} else {
+						$ship_result = $shipsGo->PostContainerInfo($ContainerNumber, $ShippingLine, $email, $Referance);
+					}
+					$object->array_options['options_updatedtime'] = dol_now();
+
+					$updatesql = "UPDATE ".MAIN_DB_PREFIX."expedition_extrafields SET";
+					$updatesql .= " requestid = ".$ship_result['RequestId'];
+					$updatesql .= ", updatedtime = '".date('Y-m-d H:i:s', $object->array_options['options_updatedtime'])."'";
+					$updatesql .= " WHERE fk_object = ".$object->id;
+					$db->query($updatesql);
+				}
+			} else if ($object->array_options['options_sailingstatusid'] != 3 && $object->array_options['options_sailingstatusid'] != 4) {
+				$ship_status = $shipsGo->GetContainerInfo($ContainerNumber)[0];	
+				if ($ship_status['Message'] == 'Success') {
+					$object->array_options['options_sailingstatusid']  = $ship_status['SailingStatusId'];
+					$object->array_options['options_pol'] = $ship_status['Pol'];
+					$object->array_options['options_atd'] = strtotime(str_replace('/', '-', $ship_status['DepartureDate']));
+					$object->array_options['options_pod'] = $ship_status['Pod'];
+					$object->array_options['options_ata'] = strtotime(str_replace('/', '-', $ship_status['ArrivalDate']));
+					$object->array_options['options_livemapurl'] = $ship_status['LiveMapUrl'];
+					$object->array_options['options_updatedtime'] = dol_now();
+
+					$updatesql = "UPDATE ".MAIN_DB_PREFIX."expedition_extrafields SET";
+					$updatesql .= " sailingstatusid = ".$ship_status['SailingStatusId'];
+					$updatesql .= ", pol = '".$ship_status['Pol']."'";
+					if (!empty($ship_status['DepartureDate'])) {
+						$updatesql .= ", atd = '".date('Y-m-d', strtotime(str_replace('/', '-', $ship_status['DepartureDate'])))."'";
+					}
+					$updatesql .= ", pod = '".$ship_status['Pod']."'";
+					if (!empty($ship_status['ArrivalDate'])) {
+						$updatesql .= ", ata = '".date('Y-m-d', strtotime(str_replace('/', '-', $ship_status['ArrivalDate'])))."'";						
+					}
+					$updatesql .= ", livemapurl = '".$ship_status['LiveMapUrl']."'";
+					$updatesql .= ", updatedtime = '".date('Y-m-d H:i:s', $object->array_options['options_updatedtime'])."'";
+					$updatesql .= " WHERE fk_object = ".$object->id;
+					$db->query($updatesql);
+				}
+			}
+		}
 
 		// Call Hook formConfirm
 		$parameters = array('formConfirm' => $formconfirm);
@@ -1796,7 +1865,7 @@ if ($action == 'create') {
 		print $langs->trans('DateDeliveryPlanned');
 		print '</td>';
 
-		if ($action != 'editdate_livraison') {
+		if ($action != 'editdate_livraison' && $object->statut == 0) {
 			print '<td class="right"><a class="editfielda" href="'.$_SERVER["PHP_SELF"].'?action=editdate_livraison&amp;id='.$object->id.'">'.img_edit($langs->trans('SetDeliveryDate'), 1).'</a></td>';
 		}
 		print '</tr></table>';
@@ -1816,7 +1885,7 @@ if ($action == 'create') {
 
 		// Weight
 		print '<tr><td>';
-		print $form->editfieldkey("Weight", 'trueWeight', $object->trueWeight, $object, $user->rights->expedition->creer);
+		print $form->editfieldkey("Weight", 'trueWeight', $object->trueWeight, $object, ($user->rights->expedition->creer && $object->statut == 0));
 		print '</td><td colspan="3">';
 
 		if ($action == 'edittrueWeight') {
@@ -1847,13 +1916,13 @@ if ($action == 'create') {
 		print '</td></tr>';
 
 		// Width
-		print '<tr><td>'.$form->editfieldkey("Width", 'trueWidth', $object->trueWidth, $object, $user->rights->expedition->creer).'</td><td colspan="3">';
+		print '<tr><td>'.$form->editfieldkey("Width", 'trueWidth', $object->trueWidth, $object, $user->rights->expedition->creer && $object->statut == 0).'</td><td colspan="3">';
 		print $form->editfieldval("Width", 'trueWidth', $object->trueWidth, $object, $user->rights->expedition->creer);
 		print ($object->trueWidth && $object->width_units != '') ? ' '.measuringUnitString(0, "size", $object->width_units) : '';
 		print '</td></tr>';
 
 		// Height
-		print '<tr><td>'.$form->editfieldkey("Height", 'trueHeight', $object->trueHeight, $object, $user->rights->expedition->creer).'</td><td colspan="3">';
+		print '<tr><td>'.$form->editfieldkey("Height", 'trueHeight', $object->trueHeight, $object, $user->rights->expedition->creer && $object->statut == 0).'</td><td colspan="3">';
 		if ($action == 'edittrueHeight') {
 			print '<form name="settrueHeight" action="'.$_SERVER["PHP_SELF"].'" method="post">';
 			print '<input name="action" value="settrueHeight" type="hidden">';
@@ -1872,7 +1941,7 @@ if ($action == 'create') {
 		print '</td></tr>';
 
 		// Depth
-		print '<tr><td>'.$form->editfieldkey("Depth", 'trueDepth', $object->trueDepth, $object, $user->rights->expedition->creer).'</td><td colspan="3">';
+		print '<tr><td>'.$form->editfieldkey("Depth", 'trueDepth', $object->trueDepth, $object, $user->rights->expedition->creer && $object->statut == 0).'</td><td colspan="3">';
 		print $form->editfieldval("Depth", 'trueDepth', $object->trueDepth, $object, $user->rights->expedition->creer);
 		print ($object->trueDepth && $object->depth_units != '') ? ' '.measuringUnitString(0, "size", $object->depth_units) : '';
 		print '</td></tr>';
@@ -1909,26 +1978,13 @@ if ($action == 'create') {
 		print "</td>\n";
 		print '</tr>';
 
-		// Other attributes
-		$cols = 2;
-		include DOL_DOCUMENT_ROOT.'/core/tpl/extrafields_view.tpl.php';
-
-		print '</table>';
-
-		print '</div>';
-		print '<div class="fichehalfright">';
-		print '<div class="ficheaddleft">';
-		print '<div class="underbanner clearboth"></div>';
-
-		print '<table class="border centpercent tableforfield">';
-
 		// Sending method
 		print '<tr><td height="10">';
 		print '<table class="nobordernopadding" width="100%"><tr><td>';
 		print $langs->trans('SendingMethod');
 		print '</td>';
 
-		if ($action != 'editshipping_method_id') {
+		if ($action != 'editshipping_method_id' && $object->statut == 0) {
 			print '<td class="right"><a class="editfielda" href="'.$_SERVER["PHP_SELF"].'?action=editshipping_method_id&amp;id='.$object->id.'">'.img_edit($langs->trans('SetSendingMethod'), 1).'</a></td>';
 		}
 		print '</tr></table>';
@@ -1955,7 +2011,7 @@ if ($action == 'create') {
 		print '</tr>';
 
 		// Tracking Number
-		print '<tr><td class="titlefield">'.$form->editfieldkey("TrackingNumber", 'tracking_number', $object->tracking_number, $object, $user->rights->expedition->creer).'</td><td colspan="3">';
+		print '<tr><td class="titlefield">'.$form->editfieldkey("TrackingNumber", 'tracking_number', $object->tracking_number, $object, ($user->rights->expedition->creer && $object->statut == 0)).'</td><td colspan="3">';
 		print $form->editfieldval("TrackingNumber", 'tracking_number', $object->tracking_url, $object, $user->rights->expedition->creer, 'safehtmlstring', $object->tracking_number);
 		print '</td></tr>';
 
@@ -1965,7 +2021,7 @@ if ($action == 'create') {
 			print '<table width="100%" class="nobordernopadding"><tr><td>';
 			print $langs->trans('IncotermLabel');
 			print '<td><td class="right">';
-			if ($user->rights->expedition->creer) {
+			if ($user->rights->expedition->creer && $object->statut == 0) {
 				print '<a class="editfielda" href="'.DOL_URL_ROOT.'/expedition/card.php?id='.$object->id.'&action=editincoterm">'.img_edit().'</a>';
 			} else {
 				print '&nbsp;';
@@ -1980,6 +2036,25 @@ if ($action == 'create') {
 			}
 			print '</td></tr>';
 		}
+		print '</table>';
+
+		// Update ShipsGo Status
+		if ($object->statut > 0 && $user->rights->expedition->creer && $user->rights->expedition->shipping_advance->validate) {
+			print '<div class="tabsAction">';
+			print '<a class="butAction" href="'.$_SERVER["PHP_SELF"].'?id='.$object->id.'&amp;action=updateships">'.$langs->trans("UpdateShips").'</a>';
+			print '</div>';
+		}
+		print '</div>';
+
+		print '<div class="fichehalfright">';
+		print '<div class="ficheaddleft">';
+		print '<div class="underbanner clearboth"></div>';
+
+		print '<table class="border centpercent tableforfield">';
+
+		// Other attributes
+		$cols = 2;
+		include DOL_DOCUMENT_ROOT.'/core/tpl/extrafields_view.tpl.php';
 
 		// Other attributes
 		$parameters = array('colspan' => ' colspan="3"', 'cols' => '3');
@@ -2086,7 +2161,7 @@ if ($action == 'create') {
 		// Get list of products already sent for same source object into $alreadysent
 		$alreadysent = array();
 		if ($origin && $origin_id > 0) {
-			$sql = "SELECT obj.rowid, obj.fk_product, obj.label, obj.description, obj.product_type as fk_product_type, obj.qty as qty_asked, obj.date_start, obj.date_end";
+			$sql = "SELECT obj.rowid, obj.fk_product, obj.label, obj.description, obj.product_type as fk_product_type, obj.qty as qty_asked, obj.fk_unit, obj.date_start, obj.date_end";
 			$sql .= ", ed.rowid as shipmentline_id, ed.qty as qty_shipped, ed.fk_expedition as expedition_id, ed.fk_origin_line, ed.fk_entrepot";
 			$sql .= ", e.rowid as shipment_id, e.ref as shipment_ref, e.date_creation, e.date_valid, e.date_delivery, e.date_expedition";
 			//if ($conf->delivery_note->enabled) $sql .= ", l.rowid as livraison_id, l.ref as livraison_ref, l.date_delivery, ld.qty as qty_received";
@@ -2164,7 +2239,7 @@ if ($action == 'create') {
 					$product_static->status = $lines[$i]->product_tosell;
 					$product_static->status_buy = $lines[$i]->product_tobuy;
 					$product_static->status_batch = $lines[$i]->product_tobatch;
-
+					
 					$product_static->weight = $lines[$i]->weight;
 					$product_static->weight_units = $lines[$i]->weight_units;
 					$product_static->length = $lines[$i]->length;
@@ -2206,8 +2281,13 @@ if ($action == 'create') {
 					print "</td>\n";
 				}
 
+				$unit_order = '';
+				if ($conf->global->PRODUCT_USE_UNITS) {
+					$unit_order = measuringUnitString($lines[$i]->fk_unit);
+				}
+
 				// Qty ordered
-				print '<td class="center linecolqty">'.$lines[$i]->qty_asked.'</td>';
+				print '<td class="center linecolqty">'.$lines[$i]->qty_asked.' '.$unit_order.'</td>';
 
 				// Qty in other shipments (with shipment and warehouse used)
 				if ($origin && $origin_id > 0) {
@@ -2226,7 +2306,7 @@ if ($action == 'create') {
 								}
 								$shipment_static->fetch($shipmentline_var['shipment_id']);
 								print $shipment_static->getNomUrl(1);
-								print ' - '.$shipmentline_var['qty_shipped'];
+								print ' - '.$shipmentline_var['qty_shipped'].' '.$unit_order;
 								$htmltext = $langs->trans("DateValidation").' : '.(empty($shipmentline_var['date_valid']) ? $langs->trans("Draft") : dol_print_date($shipmentline_var['date_valid'], 'dayhour'));
 								if (!empty($conf->stock->enabled) && $shipmentline_var['warehouse'] > 0) {
 									$warehousestatic->fetch($shipmentline_var['warehouse']);
@@ -2248,7 +2328,7 @@ if ($action == 'create') {
 						foreach ($lines[$i]->detail_batch as $detail_batch) {
 							print '<tr>';
 							// Qty to ship or shipped
-							print '<td><input class="qtyl" name="qtyl'.$detail_batch->fk_expeditiondet.'_'.$detail_batch->id.'" id="qtyl'.$line_id.'_'.$detail_batch->id.'" type="text" size="4" value="'.$detail_batch->qty.'"></td>';
+							print '<td><input class="qtyl" name="qtyl'.$detail_batch->fk_expeditiondet.'_'.$detail_batch->id.'" id="qtyl'.$line_id.'_'.$detail_batch->id.'" type="text" size="4" value="'.$detail_batch->qty.'">' .$unit_order.'</td>';
 							// Batch number managment
 							if ($lines[$i]->entrepot_id == 0) {
 								// only show lot numbers from src warehouse when shipping from multiple warehouses
@@ -2297,7 +2377,7 @@ if ($action == 'create') {
 							print '<!-- case edit 5 -->';
 							print '<tr>';
 							// Qty to ship or shipped
-							print '<td><input class="qtyl" name="qtyl'.$line_id.'" id="qtyl'.$line_id.'" type="text" size="4" value="'.$lines[$i]->qty_shipped.'"></td>';
+							print '<td><input class="qtyl" name="qtyl'.$line_id.'" id="qtyl'.$line_id.'" type="text" size="4" value="'.$lines[$i]->qty_shipped.'">' .$unit_order.'</td>';
 							// Warehouse source
 							print '<td></td>';
 							// Batch number managment
@@ -2308,7 +2388,7 @@ if ($action == 'create') {
 						print '<!-- case edit 6 -->';
 						print '<tr>';
 						// Qty to ship or shipped
-						print '<td><input class="qtyl" name="qtyl'.$line_id.'" id="qtyl'.$line_id.'" type="text" size="4" value="'.$lines[$i]->qty_shipped.'"></td>';
+						print '<td><input class="qtyl" name="qtyl'.$line_id.'" id="qtyl'.$line_id.'" type="text" size="4" value="'.$lines[$i]->qty_shipped.'">' .$unit_order.'</td>';
 						// Warehouse source
 						print '<td></td>';
 						// Batch number managment
@@ -2319,7 +2399,7 @@ if ($action == 'create') {
 					print '</table></td>';
 				} else {
 					// Qty to ship or shipped
-					print '<td class="linecolqtytoship center">'.$lines[$i]->qty_shipped.'</td>';
+					print '<td class="linecolqtytoship center">'.$lines[$i]->qty_shipped.' '.$unit_order.'</td>';
 
 					// Warehouse source
 					if (!empty($conf->stock->enabled)) {
@@ -2477,7 +2557,7 @@ if ($action == 'create') {
 
 			// TODO add alternative status
 			// 0=draft, 1=validated, 2=billed, we miss a status "delivered" (only available on order)
-			if ($object->statut == Expedition::STATUS_CLOSED && $user->rights->expedition->creer) {
+			if ($object->statut == Expedition::STATUS_CLOSED && $object->billed && $user->rights->expedition->creer) {
 				if (!empty($conf->facture->enabled) && !empty($conf->global->WORKFLOW_BILL_ON_SHIPMENT)) {  // Quand l'option est on, il faut avoir le bouton en plus et non en remplacement du Close ?
 					print '<a class="butAction" href="'.$_SERVER["PHP_SELF"].'?id='.$object->id.'&amp;action=reopen">'.$langs->trans("ClassifyUnbilled").'</a>';
 				} else {
@@ -2485,6 +2565,11 @@ if ($action == 'create') {
 				}
 			}
 
+			// Modify
+			if (($object->statut == Expedition::STATUS_VALIDATED || ($object->statut == Expedition::STATUS_CLOSED && !$object->billed)) && $user->rights->expedition->creer && $user->rights->expedition->shipping_advance->validate) {
+				print '<a class="butAction" href="'.$_SERVER["PHP_SELF"].'?id='.$object->id.'&amp;action=modif">'.$langs->trans("Modify").'</a>';
+			}
+
 			// Send
 			if (empty($user->socid)) {
 				if ($object->statut > 0) {
@@ -2511,11 +2596,11 @@ if ($action == 'create') {
 				print '<a class="butAction" href="'.$_SERVER["PHP_SELF"].'?id='.$object->id.'&amp;action=create_delivery">'.$langs->trans("CreateDeliveryOrder").'</a>';
 			}
 			// Close
-			if ($object->statut == Expedition::STATUS_VALIDATED) {
-				if ($user->rights->expedition->creer && $object->statut > 0 && !$object->billed) {
+			if ($object->statut == Expedition::STATUS_VALIDATED || ($object->statut == Expedition::STATUS_CLOSED && !$object->billed)) {
+				if ($user->rights->expedition->creer && $object->statut > 0 && (($object->statut == Expedition::STATUS_VALIDATED && $object->billed) || !$object->billed)) {
 					$label = "Close"; $paramaction = 'classifyclosed'; // = Transferred/Received
 					// Label here should be "Close" or "ClassifyBilled" if we decided to make bill on shipments instead of orders
-					if (!empty($conf->facture->enabled) && !empty($conf->global->WORKFLOW_BILL_ON_SHIPMENT)) {  // Quand l'option est on, il faut avoir le bouton en plus et non en remplacement du Close ?
+					if (!empty($conf->facture->enabled) && !empty($conf->global->WORKFLOW_BILL_ON_SHIPMENT) && !$object->billed) {  // Quand l'option est on, il faut avoir le bouton en plus et non en remplacement du Close ?
 						$label = "ClassifyBilled";
 						$paramaction = 'classifybilled';
 					}
