diff --git a/htdocs/comm/action/card.php b/htdocs/comm/action/card.php
index 681a977adb1..752868acb87 100644
--- a/htdocs/comm/action/card.php
+++ b/htdocs/comm/action/card.php
@@ -659,7 +659,7 @@ if (empty($reshook) && $action == 'update') {
 				$categories = GETPOST('categories', 'array');
 				$object->setCategories($categories);
 
-				$object->loadReminders($remindertype, 0, false);
+				$object->loadReminders();
 				if (!empty($object->reminders) && $object->datep > dol_now()) {
 					foreach ($object->reminders as $reminder) {
 						$reminder->delete($user);
@@ -756,7 +756,7 @@ if (empty($reshook) && GETPOST('actionmove', 'alpha') == 'mupdate') {
 
 	$newdate = GETPOST('newdate', 'alpha');
 	if (empty($newdate) || strpos($newdate, 'dayevent_') != 0) {
-		header("Location: ".$backtopage, true, 307);
+		header("Location: ".$backtopage);
 		exit;
 	}
 
@@ -841,7 +841,7 @@ if (empty($reshook) && GETPOST('actionmove', 'alpha') == 'mupdate') {
 		}
 	}
 	if (!empty($backtopage)) {
-		header("Location: ".$backtopage, true, 307);
+		header("Location: ".$backtopage);
 		exit;
 	} else {
 		$action = '';
@@ -1181,14 +1181,8 @@ if ($action == 'create') {
 			$preselectedids[GETPOST('contactid', 'int')] = GETPOST('contactid', 'int');
 		}
 		if ($origin=='contact') $preselectedids[GETPOST('originid', 'int')] = GETPOST('originid', 'int');
-		// select "all" or "none" contact by default
-		if (getDolGlobalInt('MAIN_ACTIONCOM_CAN_ADD_ANY_CONTACT')) {
-			$select_contact_default = 0; // select "all" contacts by default : avoid to use it if there is a lot of contacts
-		} else {
-			$select_contact_default = -1; // select "none" by default
-		}
 		print img_picto('', 'contact', 'class="paddingrightonly"');
-		print $form->selectcontacts(GETPOSTISSET('socid') ? GETPOSTINT('socid') : $select_contact_default, $preselectedids, 'socpeopleassigned[]', 1, '', '', 0, 'minwidth300 quatrevingtpercent', false, 0, array(), false, 'multiple', 'contactid');
+		print $form->selectcontacts(GETPOST('socid', 'int'), $preselectedids, 'socpeopleassigned[]', 1, '', '', 0, 'minwidth300 quatrevingtpercent', false, 0, array(), false, 'multiple', 'contactid');
 		print '</td></tr>';
 	}
 
@@ -1445,10 +1439,6 @@ if ($id > 0) {
 	            		$("#fullday").change(function() {
 	            			setdatefields();
 	            		});
-	            		$("#actioncode").change(function() {
-                        	if ($("#actioncode").val() == \'AC_RDV\') $("#dateend").addClass("fieldrequired");
-                        	else $("#dateend").removeClass("fieldrequired");
-                    	});
                    })';
 			print '</script>'."\n";
 		}
@@ -1493,12 +1483,7 @@ if ($id > 0) {
 		print '<tr><td>'.$langs->trans("EventOnFullDay").'</td><td colspan="3"><input type="checkbox" id="fullday" name="fullday" '.($object->fulldayevent ? ' checked' : '').'></td></tr>';
 
 		// Date start - end
-		print '<tr><td class="nowrap">';
-		print '<span class="fieldrequired">'.$langs->trans("DateActionStart").'</span>';
-		print ' - ';
-		print '<span id="dateend"'.($object->type_code == 'AC_RDV' ? ' class="fieldrequired"' : '').'>'.$langs->trans("DateActionEnd").'</span>';
-		print '</td><td td colspan="3">';
-		//print '<tr><td class="nowrap"><span class="fieldrequired">'.$langs->trans("DateActionStart").' - '.$langs->trans("DateActionEnd").'</span></td><td colspan="3">';
+		print '<tr><td class="nowrap"><span class="fieldrequired">'.$langs->trans("DateActionStart").' - '.$langs->trans("DateActionEnd").'</span></td><td colspan="3">';
 		if (GETPOST("afaire") == 1) {
 			print $form->selectDate($datep ? $datep : $object->datep, 'ap', 1, 1, 0, "action", 1, 1, 0, 'fulldaystart', '', '', '', 1, '', '', 'tzuser');
 		} elseif (GETPOST("afaire") == 2) {
@@ -1670,12 +1655,7 @@ if ($id > 0) {
 			// related contact
 			print '<tr><td>'.$langs->trans("ActionOnContact").'</td><td>';
 			print '<div class="maxwidth200onsmartphone">';
-			if (getDolGlobalInt('MAIN_ACTIONCOM_CAN_ADD_ANY_CONTACT')) {
-				$select_contact_default = 0; // select "all" contacts by default : avoid to use it if there is a lot of contacts
-			} else {
-				$select_contact_default = -1; // select "none" by default
-			}
-			print img_picto('', 'contact', 'class="paddingrightonly"').$form->selectcontacts(!empty($object->socid) ? $object->socid : $select_contact_default, array_keys($object->socpeopleassigned), 'socpeopleassigned[]', 1, '', '', 1, 'quatrevingtpercent', false, 0, 0, array(), 'multiple', 'contactid');
+			print img_picto('', 'contact', 'class="paddingrightonly"').$form->selectcontacts($object->socid, array_keys($object->socpeopleassigned), 'socpeopleassigned[]', 1, '', '', 1, 'quatrevingtpercent', false, 0, 0, array(), 'multiple', 'contactid');
 			print '</div>';
 			print '</td>';
 			print '</tr>';
diff --git a/htdocs/comm/action/class/actioncomm.class.php b/htdocs/comm/action/class/actioncomm.class.php
index 9e1e471bbd3..769bd9ddd50 100644
--- a/htdocs/comm/action/class/actioncomm.class.php
+++ b/htdocs/comm/action/class/actioncomm.class.php
@@ -708,8 +708,6 @@ class ActionComm extends CommonObject
 				$action = '';
 				$reshook = $hookmanager->executeHooks('createFrom', $parameters, $this, $action); // Note that $action and $object may have been modified by some hooks
 				if ($reshook < 0) {
-					$this->errors += $hookmanager->errors;
-					$this->error = $hookmanager->error;
 					$error++;
 				}
 			}
@@ -1203,14 +1201,11 @@ class ActionComm extends CommonObject
 
 				if (!empty($this->socpeopleassigned)) {
 					$already_inserted = array();
-					foreach (array_keys($this->socpeopleassigned) as $key => $val) {
-						if (!is_array($val)) {	// For backward compatibility when val=id
-							$val = array('id'=>$val);
-						}
+					foreach (array_keys($this->socpeopleassigned) as $id) {
 						if (!empty($already_inserted[$val['id']])) continue;
 
 						$sql = "INSERT INTO ".MAIN_DB_PREFIX."actioncomm_resources(fk_actioncomm, element_type, fk_element, mandatory, transparency, answer_status)";
-						$sql .= " VALUES(".((int) $this->id).", 'socpeople', ".((int) $val['id']).", 0, 0, 0)";
+						$sql .= " VALUES(".$this->id.", 'socpeople', ".$id.", 0, 0, 0)";
 
 						$resql = $this->db->query($sql);
 						if (!$resql) {
@@ -2344,7 +2339,6 @@ class ActionComm extends CommonObject
 
 					// Load event
 					$res = $this->fetch($actionCommReminder->fk_actioncomm);
-					if ($res > 0) $res = $this->fetch_thirdparty();
 					if ($res > 0) {
 						// PREPARE EMAIL
 						$errormesg = '';
diff --git a/htdocs/comm/action/index.php b/htdocs/comm/action/index.php
index 51d0f99b6b8..722423c0cd9 100644
--- a/htdocs/comm/action/index.php
+++ b/htdocs/comm/action/index.php
@@ -979,8 +979,8 @@ if ($resql) {
 		$event->type = 'holiday';
 		$event->type_picto = 'holiday';
 
-		$event->datep                   = $db->jdate($obj->date_start) + (empty($halfday) || $halfday == 1 ? 0 : 12 * 60 * 60 - 1);
-		$event->datef                   = $db->jdate($obj->date_end) + (empty($halfday) || $halfday == -1 ? 24 : 12) * 60 * 60 - 1;
+		$event->datep                   = $db->jdate($obj->date_start);
+		$event->datef                   = $db->jdate($obj->date_end);
 		$event->date_start_in_calendar  = $event->datep;
 		$event->date_end_in_calendar    = $event->datef;
 
diff --git a/htdocs/comm/index.php b/htdocs/comm/index.php
index 18546332ed0..318401ad136 100644
--- a/htdocs/comm/index.php
+++ b/htdocs/comm/index.php
@@ -122,6 +122,9 @@ if ($tmp) {
  */
 if (!empty($conf->propal->enabled) && $user->rights->propal->lire) {
 	$sql = "SELECT p.rowid, p.ref, p.ref_client, p.total_ht, p.total_tva, p.total_ttc, p.fk_statut as status";
+	if (!empty($conf->multicurrency->enabled)) {
+		$sql .= ", p.multicurrency_total_ht, p.multicurrency_total_tva, p.multicurrency_total_ttc, p.multicurrency_code";
+	}
 	$sql .= ", s.rowid as socid, s.nom as name, s.name_alias";
 	$sql .= ", s.code_client, s.code_compta, s.client";
 	$sql .= ", s.code_fournisseur, s.code_compta_fournisseur, s.fournisseur";
@@ -188,7 +191,11 @@ if (!empty($conf->propal->enabled) && $user->rights->propal->lire) {
 				print '<tr class="oddeven">';
 				print '<td class="nowrap tdoverflowmax100">'.$propalstatic->getNomUrl(1).'</td>';
 				print '<td class="nowrap tdoverflowmax100">'.$companystatic->getNomUrl(1, 'customer').'</td>';
-				print '<td class="nowrap right tdamount amount">'.price((!empty($conf->global->MAIN_DASHBOARD_USE_TOTAL_HT) ? $obj->total_ht : $obj->total_ttc)).'</td>';
+				if (!empty($obj->multicurrency_code)) {
+					print '<td class="nowrap right tdamount amount">'.price((!empty($conf->global->MAIN_DASHBOARD_USE_TOTAL_HT) ? $obj->multicurrency_total_ht : $obj->multicurrency_total_ttc), 0, '', 1, -1, -1, $obj->multicurrency_code).'</td>';
+				} else {
+					print '<td class="nowrap right tdamount amount">'.price((!empty($conf->global->MAIN_DASHBOARD_USE_TOTAL_HT) ? $obj->total_ht : $obj->total_ttc), 0, '', 1, -1, -1, $conf->currency).'</td>';
+				}
 				print '</tr>';
 
 				$i++;
@@ -219,6 +226,9 @@ if (!empty($conf->propal->enabled) && $user->rights->propal->lire) {
  */
 if (!empty($conf->supplier_proposal->enabled) && $user->rights->supplier_proposal->lire) {
 	$sql = "SELECT p.rowid, p.ref, p.total_ht, p.total_tva, p.total_ttc, p.fk_statut as status";
+	if (!empty($conf->multicurrency->enabled)) {
+		$sql .= ", p.multicurrency_total_ht, p.multicurrency_total_tva, p.multicurrency_total_ttc, p.multicurrency_code";
+	}
 	$sql .= ", s.rowid as socid, s.nom as name, s.name_alias";
 	$sql .= ", s.code_client, s.code_compta, s.client";
 	$sql .= ", s.code_fournisseur, s.code_compta_fournisseur, s.fournisseur";
@@ -284,7 +294,11 @@ if (!empty($conf->supplier_proposal->enabled) && $user->rights->supplier_proposa
 				print '<tr class="oddeven">';
 				print '<td class="nowrap tdoverflowmax100">'.$supplierproposalstatic->getNomUrl(1).'</td>';
 				print '<td class="nowrap tdoverflowmax100">'.$companystatic->getNomUrl(1, 'supplier').'</td>';
-				print '<td class="nowrap right tdamount amount">'.price(!empty($conf->global->MAIN_DASHBOARD_USE_TOTAL_HT) ? $obj->total_ht : $obj->total_ttc).'</td>';
+				if (!empty($obj->multicurrency_code)) {
+					print '<td class="nowrap right tdamount amount">'.price((!empty($conf->global->MAIN_DASHBOARD_USE_TOTAL_HT) ? $obj->multicurrency_total_ht : $obj->multicurrency_total_ttc), 0, '', 1, -1, -1, $obj->multicurrency_code).'</td>';
+				} else {
+					print '<td class="nowrap right tdamount amount">'.price((!empty($conf->global->MAIN_DASHBOARD_USE_TOTAL_HT) ? $obj->total_ht : $obj->total_ttc), 0, '', 1, -1, -1, $conf->currency).'</td>';
+				}
 				print '</tr>';
 
 				$i++;
@@ -315,6 +329,9 @@ if (!empty($conf->supplier_proposal->enabled) && $user->rights->supplier_proposa
  */
 if (!empty($conf->commande->enabled) && $user->rights->commande->lire) {
 	$sql = "SELECT c.rowid, c.ref, c.ref_client, c.total_ht, c.total_tva, c.total_ttc, c.fk_statut as status";
+	if (!empty($conf->multicurrency->enabled)) {
+		$sql .= ", c.multicurrency_total_ht, c.multicurrency_total_tva, c.multicurrency_total_ttc, c.multicurrency_code";
+	}
 	$sql .= ", s.rowid as socid, s.nom as name, s.name_alias";
 	$sql .= ", s.code_client, s.code_compta, s.client";
 	$sql .= ", s.code_fournisseur, s.code_compta_fournisseur, s.fournisseur";
@@ -381,7 +398,11 @@ if (!empty($conf->commande->enabled) && $user->rights->commande->lire) {
 				print '<tr class="oddeven">';
 				print '<td class="nowrap tdoverflowmax100">'.$orderstatic->getNomUrl(1).'</td>';
 				print '<td class="nowrap tdoverflowmax100">'.$companystatic->getNomUrl(1, 'customer').'</td>';
-				print '<td class="nowrap right tdamount amount">'.price(!empty($conf->global->MAIN_DASHBOARD_USE_TOTAL_HT) ? $obj->total_ht : $obj->total_ttc).'</td>';
+				if (!empty($obj->multicurrency_code)) {
+					print '<td class="nowrap right tdamount amount">'.price((!empty($conf->global->MAIN_DASHBOARD_USE_TOTAL_HT) ? $obj->multicurrency_total_ht : $obj->multicurrency_total_ttc), 0, '', 1, -1, -1, $obj->multicurrency_code).'</td>';
+				} else {
+					print '<td class="nowrap right tdamount amount">'.price((!empty($conf->global->MAIN_DASHBOARD_USE_TOTAL_HT) ? $obj->total_ht : $obj->total_ttc), 0, '', 1, -1, -1, $conf->currency).'</td>';
+				}
 				print '</tr>';
 
 				$i++;
@@ -412,6 +433,9 @@ if (!empty($conf->commande->enabled) && $user->rights->commande->lire) {
  */
 if ((!empty($conf->fournisseur->enabled) && empty($conf->global->MAIN_USE_NEW_SUPPLIERMOD) && $user->rights->fournisseur->commande->lire) || (!empty($conf->supplier_order->enabled) && $user->rights->supplier_order->lire)) {
 	$sql = "SELECT cf.rowid, cf.ref, cf.ref_supplier, cf.total_ht, cf.total_tva, cf.total_ttc, cf.fk_statut as status";
+	if (!empty($conf->multicurrency->enabled)) {
+		$sql .= ", cf.multicurrency_total_ht, cf.multicurrency_total_tva, cf.multicurrency_total_ttc, cf.multicurrency_code";
+	}
 	$sql .= ", s.rowid as socid, s.nom as name, s.name_alias";
 	$sql .= ", s.code_client, s.code_compta, s.client";
 	$sql .= ", s.code_fournisseur, s.code_compta_fournisseur, s.fournisseur";
@@ -478,7 +502,11 @@ if ((!empty($conf->fournisseur->enabled) && empty($conf->global->MAIN_USE_NEW_SU
 				print '<tr class="oddeven">';
 				print '<td class="nowrap tdoverflowmax100">'.$supplierorderstatic->getNomUrl(1).'</td>';
 				print '<td class="nowrap tdoverflowmax100">'.$companystatic->getNomUrl(1, 'supplier').'</td>';
-				print '<td class="nowrap right tdamount amount">'.price(!empty($conf->global->MAIN_DASHBOARD_USE_TOTAL_HT) ? $obj->total_ht : $obj->total_ttc).'</td>';
+				if (!empty($obj->multicurrency_code)) {
+					print '<td class="nowrap right tdamount amount">'.price((!empty($conf->global->MAIN_DASHBOARD_USE_TOTAL_HT) ? $obj->multicurrency_total_ht : $obj->multicurrency_total_ttc), 0, '', 1, -1, -1, $obj->multicurrency_code).'</td>';
+				} else {
+					print '<td class="nowrap right tdamount amount">'.price((!empty($conf->global->MAIN_DASHBOARD_USE_TOTAL_HT) ? $obj->total_ht : $obj->total_ttc), 0, '', 1, -1, -1, $conf->currency).'</td>';
+				}
 				print '</tr>';
 
 				$i++;
@@ -786,6 +814,9 @@ if (!empty($conf->contrat->enabled) && $user->rights->contrat->lire && 0) { // T
  */
 if (!empty($conf->propal->enabled) && $user->rights->propal->lire) {
 	$sql = "SELECT p.rowid as propalid, p.entity, p.total_ttc, p.total_ht, p.total_tva, p.ref, p.ref_client, p.fk_statut, p.datep as dp, p.fin_validite as dfv";
+	if (!empty($conf->multicurrency->enabled)) {
+		$sql .= ", p.multicurrency_total_ht, p.multicurrency_total_tva, p.multicurrency_total_ttc, p.multicurrency_code";
+	}
 	$sql .= ", s.rowid as socid, s.nom as name, s.name_alias";
 	$sql .= ", s.code_client, s.code_compta, s.client";
 	$sql .= ", s.code_fournisseur, s.code_compta_fournisseur, s.fournisseur";
@@ -868,7 +899,11 @@ if (!empty($conf->propal->enabled) && $user->rights->propal->lire) {
 
 				print '<td class="nowrap">'.$companystatic->getNomUrl(1, 'customer', 44).'</td>';
 				print '<td class="right tddate">'.dol_print_date($db->jdate($obj->dp), 'day').'</td>';
-				print '<td class="right tdamount amount">'.price(!empty($conf->global->MAIN_DASHBOARD_USE_TOTAL_HT) ? $obj->total_ht : $obj->total_ttc).'</td>';
+				if (!empty($obj->multicurrency_code)) {
+					print '<td class="nowrap right tdamount amount">'.price((!empty($conf->global->MAIN_DASHBOARD_USE_TOTAL_HT) ? $obj->multicurrency_total_ht : $obj->multicurrency_total_ttc), 0, '', 1, -1, -1, $obj->multicurrency_code).'</td>';
+				} else {
+					print '<td class="nowrap right tdamount amount">'.price((!empty($conf->global->MAIN_DASHBOARD_USE_TOTAL_HT) ? $obj->total_ht : $obj->total_ttc), 0, '', 1, -1, -1, $conf->currency).'</td>';
+				}
 				print '<td align="center" width="14">'.$propalstatic->LibStatut($obj->fk_statut, 3).'</td>';
 
 				print '</tr>';
@@ -902,6 +937,9 @@ if (!empty($conf->propal->enabled) && $user->rights->propal->lire) {
  */
 if (!empty($conf->commande->enabled) && $user->rights->commande->lire) {
 	$sql = "SELECT c.rowid as commandeid, c.total_ttc, c.total_ht, c.total_tva, c.ref, c.ref_client, c.fk_statut, c.date_valid as dv, c.facture as billed";
+	if (!empty($conf->multicurrency->enabled)) {
+		$sql .= ", c.multicurrency_total_ht, c.multicurrency_total_tva, c.multicurrency_total_ttc, c.multicurrency_code";
+	}
 	$sql .= ", s.rowid as socid, s.nom as name, s.name_alias";
 	$sql .= ", s.code_client, s.code_compta, s.client";
 	$sql .= ", s.code_fournisseur, s.code_compta_fournisseur, s.fournisseur";
@@ -985,7 +1023,11 @@ if (!empty($conf->commande->enabled) && $user->rights->commande->lire) {
 
 				print '<td class="nowrap">'.$companystatic->getNomUrl(1, 'customer', 44).'</td>';
 				print '<td class="right tddate">'.dol_print_date($db->jdate($obj->dv), 'day').'</td>';
-				print '<td class="right tdamount amount">'.price(!empty($conf->global->MAIN_DASHBOARD_USE_TOTAL_HT) ? $obj->total_ht : $obj->total_ttc).'</td>';
+				if (!empty($obj->multicurrency_code)) {
+					print '<td class="nowrap right tdamount amount">'.price((!empty($conf->global->MAIN_DASHBOARD_USE_TOTAL_HT) ? $obj->multicurrency_total_ht : $obj->multicurrency_total_ttc), 0, '', 1, -1, -1, $obj->multicurrency_code).'</td>';
+				} else {
+					print '<td class="nowrap right tdamount amount">'.price((!empty($conf->global->MAIN_DASHBOARD_USE_TOTAL_HT) ? $obj->total_ht : $obj->total_ttc), 0, '', 1, -1, -1, $conf->currency).'</td>';
+				}
 				print '<td align="center" width="14">'.$orderstatic->LibStatut($obj->fk_statut, $obj->billed, 3).'</td>';
 
 				print '</tr>';
diff --git a/htdocs/comm/mailing/card.php b/htdocs/comm/mailing/card.php
index b8bd18f56c2..b91bc866d9a 100644
--- a/htdocs/comm/mailing/card.php
+++ b/htdocs/comm/mailing/card.php
@@ -508,10 +508,9 @@ if (empty($reshook)) {
 				exit;
 			}
 			$mesgs[] = $object->error;
-			$mesgs = array_merge($mesgs, $object->errors);
 		}
 
-		setEventMessages('', $mesgs, 'errors');
+		setEventMessages(null, $mesgs, 'errors');
 		$action = "create";
 	}
 
@@ -596,10 +595,9 @@ if (empty($reshook)) {
 					exit;
 				}
 				$mesgs[] = $object->error;
-				$mesgs = array_merge($mesgs, $object->errors);
 			}
 
-			setEventMessages('', $mesgs, 'errors');
+			setEventMessages($mesg, $mesgs, 'errors');
 			$action = "edit";
 		} else {
 			$action = "edit";
diff --git a/htdocs/comm/propal/card.php b/htdocs/comm/propal/card.php
index 58b40838cc8..c96f24760c0 100644
--- a/htdocs/comm/propal/card.php
+++ b/htdocs/comm/propal/card.php
@@ -121,7 +121,7 @@ $usercansend = (empty($conf->global->MAIN_USE_ADVANCED_PERMS) || $user->rights->
 $usercancreateorder = $user->rights->commande->creer;
 $usercancreateinvoice = $user->rights->facture->creer;
 $usercancreatecontract = $user->rights->contrat->creer;
-$usercancreateintervention = empty($conf->ficheinter->enabled) ? 0 : $user->rights->ficheinter->creer;
+$usercancreateintervention = $user->rights->ficheinter->creer;
 $usercancreatepurchaseorder = ($user->rights->fournisseur->commande->creer || $user->rights->supplier_order->creer);
 
 $permissionnote = $usercancreate; // Used by the include of actions_setnotes.inc.php
@@ -231,11 +231,8 @@ if (empty($reshook)) {
 		// Remove line
 		$result = $object->deleteline($lineid);
 		// reorder lines
-		if ($result > 0) {
+		if ($result) {
 			$object->line_order(true);
-		} else {
-			$langs->load("errors");
-			setEventMessages($object->error, $object->errors, 'errors');
 		}
 
 		if (empty($conf->global->MAIN_DISABLE_PDF_AUTOUPDATE)) {
@@ -567,7 +564,6 @@ if (empty($reshook)) {
 							$reshook = $hookmanager->executeHooks('createFrom', $parameters, $object, $action); // Note that $action and $object may have been
 																											   // modified by hook
 							if ($reshook < 0) {
-								setEventMessages($hookmanager->error, $hookmanager->errors, 'errors');
 								$error++;
 							}
 						} else {
@@ -875,8 +871,7 @@ if (empty($reshook)) {
 			}
 		}
 
-		$propal_qty_requirement = (!empty($conf->global->PROPAL_ENABLE_NEGATIVE_QTY) ? ($qty >= 0 || $qty <= 0) : $qty >= 0);
-		if (!$error && $propal_qty_requirement && (!empty($product_desc) || (!empty($idprod) && $idprod > 0))) {
+		if (!$error && ($qty >= 0) && (!empty($product_desc) || (!empty($idprod) && $idprod > 0))) {
 			$pu_ht = 0;
 			$pu_ttc = 0;
 			$price_min = 0;
@@ -1587,7 +1582,7 @@ if ($action == 'create') {
 		//$warehouse_id       = $soc->warehouse_id;
 	} else {
 		print '<td>';
-		print img_picto('', 'company').$form->select_company('', 'socid', '((s.client = 1 OR s.client = 2 OR s.client = 3) AND status=1)', 'SelectThirdParty', 0, 0, null, 0, 'minwidth300 maxwidth500 widthcentpercentminusxx');
+		print img_picto('', 'company').$form->select_company('', 'socid', '(s.client = 1 OR s.client = 2 OR s.client = 3) AND status=1', 'SelectThirdParty', 0, 0, null, 0, 'minwidth300 maxwidth500 widthcentpercentminusxx');
 		// reload page to retrieve customer informations
 		if (empty($conf->global->RELOAD_PAGE_ON_CUSTOMER_CHANGE_DISABLED)) {
 			print '<script type="text/javascript">
@@ -1637,13 +1632,13 @@ if ($action == 'create') {
 	// Terms of payment
 	print '<tr><td class="nowrap">'.$langs->trans('PaymentConditionsShort').'</td><td>';
 	print img_picto('', 'paiment');
-	$form->select_conditions_paiements(((GETPOSTISSET('cond_reglement_id') && GETPOST('cond_reglement_id', 'int') > 0) ? GETPOST('cond_reglement_id', 'int') : $soc->cond_reglement_id), 'cond_reglement_id', -1, 1);
+	$form->select_conditions_paiements((GETPOSTISSET('cond_reglement_id') ? GETPOST('cond_reglement_id', 'int') : $soc->cond_reglement_id), 'cond_reglement_id', -1, 1);
 	print '</td></tr>';
 
 	// Mode of payment
 	print '<tr><td>'.$langs->trans('PaymentMode').'</td><td>';
 	print img_picto('', 'bank').'&ensp;';
-	$form->select_types_paiements(((GETPOSTISSET('mode_reglement_id') && GETPOST('mode_reglement_id', 'int') > 0) ? GETPOST('mode_reglement_id', 'int') : $soc->mode_reglement_id), 'mode_reglement_id', 'CRDT', 0, 1, 0, 0, 1, 'maxwidth200 widthcentpercentminusx');
+	$form->select_types_paiements((GETPOSTISSET('mode_reglement_id') ? GETPOST('mode_reglement_id', 'int') : $soc->mode_reglement_id), 'mode_reglement_id', 'CRDT', 0, 1, 0, 0, 1, 'maxwidth200 widthcentpercentminusx');
 	print '</td></tr>';
 
 	// Bank Account
@@ -1893,8 +1888,9 @@ if ($action == 'create') {
 	/*
 	 * Show object in view mode
 	 */
-	$object->fetch_thirdparty();
-	$soc = $object->thirdparty;
+
+	$soc = new Societe($db);
+	$soc->fetch($object->socid);
 
 	$head = propal_prepare_head($object);
 	print dol_get_fiche_head($head, 'comm', $langs->trans('Proposal'), -1, 'propal');
@@ -1948,7 +1944,7 @@ if ($action == 'create') {
 
 		// We verify whether the object is provisionally numbering
 		$ref = substr($object->ref, 1, 4);
-		if ($ref == 'PROV' || $ref == '') {
+		if ($ref == 'PROV') {
 			$numref = $object->getNextNumRef($soc);
 			if (empty($numref)) {
 				$error++;
@@ -1993,9 +1989,9 @@ if ($action == 'create') {
 	$morehtmlref .= $form->editfieldkey("RefCustomer", 'ref_client', $object->ref_client, $object, $usercancreate, 'string', '', 0, 1);
 	$morehtmlref .= $form->editfieldval("RefCustomer", 'ref_client', $object->ref_client, $object, $usercancreate, 'string', '', null, null, '', 1);
 	// Thirdparty
-	$morehtmlref .= '<br>'.$langs->trans('ThirdParty').' : '.$soc->getNomUrl(1, 'customer');
-	if (empty($conf->global->MAIN_DISABLE_OTHER_LINK) && $soc->id > 0) {
-		$morehtmlref .= ' (<a href="'.DOL_URL_ROOT.'/comm/propal/list.php?socid='.$soc->id.'&search_societe='.urlencode($soc->name).'">'.$langs->trans("OtherProposals").'</a>)';
+	$morehtmlref .= '<br>'.$langs->trans('ThirdParty').' : '.$object->thirdparty->getNomUrl(1, 'customer');
+	if (empty($conf->global->MAIN_DISABLE_OTHER_LINK) && $object->thirdparty->id > 0) {
+		$morehtmlref .= ' (<a href="'.DOL_URL_ROOT.'/comm/propal/list.php?socid='.$object->thirdparty->id.'&search_societe='.urlencode($object->thirdparty->name).'">'.$langs->trans("OtherProposals").'</a>)';
 	}
 	// Project
 	if (!empty($conf->projet->enabled)) {
diff --git a/htdocs/comm/propal/class/api_proposals.class.php b/htdocs/comm/propal/class/api_proposals.class.php
index 1ce7f82b99d..e4b9a266198 100644
--- a/htdocs/comm/propal/class/api_proposals.class.php
+++ b/htdocs/comm/propal/class/api_proposals.class.php
@@ -395,8 +395,8 @@ class Proposals extends DolibarrApi
 
 		$request_data = (object) $request_data;
 
-		if (isset($request_data->desc)) $request_data->desc = checkVal($request_data->desc, 'restricthtml');
-		if (isset($request_data->label)) $request_data->label = checkVal($request_data->label);
+		$request_data->desc = checkVal($request_data->desc, 'restricthtml');
+		$request_data->label = checkVal($request_data->label);
 
 		$propalline = new PropaleLigne($this->db);
 		$result = $propalline->fetch($lineid);
diff --git a/htdocs/comm/propal/class/propal.class.php b/htdocs/comm/propal/class/propal.class.php
index e09fb5f5e6e..0142d0c1b7a 100644
--- a/htdocs/comm/propal/class/propal.class.php
+++ b/htdocs/comm/propal/class/propal.class.php
@@ -183,6 +183,7 @@ class Propal extends CommonObject
 	public $fin_validite;
 
 	public $user_author_id;
+	public $user_valid_id;
 	public $user_close_id;
 
 	/**
@@ -309,8 +310,8 @@ class Propal extends CommonObject
 		'fk_currency' =>array('type'=>'varchar(3)', 'label'=>'Currency', 'enabled'=>1, 'visible'=>-1, 'position'=>155),
 		'fk_cond_reglement' =>array('type'=>'integer', 'label'=>'PaymentTerm', 'enabled'=>1, 'visible'=>-1, 'position'=>160),
 		'fk_mode_reglement' =>array('type'=>'integer', 'label'=>'PaymentMode', 'enabled'=>1, 'visible'=>-1, 'position'=>165),
-		'note_private' =>array('type'=>'text', 'label'=>'NotePrivate', 'enabled'=>1, 'visible'=>0, 'position'=>170),
-		'note_public' =>array('type'=>'text', 'label'=>'NotePublic', 'enabled'=>1, 'visible'=>0, 'position'=>175),
+		'note_private' =>array('type'=>'text', 'label'=>'NotePublic', 'enabled'=>1, 'visible'=>0, 'position'=>170),
+		'note_public' =>array('type'=>'text', 'label'=>'NotePrivate', 'enabled'=>1, 'visible'=>0, 'position'=>175),
 		'model_pdf' =>array('type'=>'varchar(255)', 'label'=>'PDFTemplate', 'enabled'=>1, 'visible'=>0, 'position'=>180),
 		'date_livraison' =>array('type'=>'date', 'label'=>'DateDeliveryPlanned', 'enabled'=>1, 'visible'=>-1, 'position'=>185),
 		'fk_shipping_method' =>array('type'=>'integer', 'label'=>'ShippingMethod', 'enabled'=>1, 'visible'=>-1, 'position'=>190),
@@ -979,8 +980,6 @@ class Propal extends CommonObject
 				$this->db->commit();
 				return 1;
 			} else {
-				$this->error = $line->error;
-				$this->errors = $line->errors;
 				$this->db->rollback();
 				return -1;
 			}
@@ -1368,7 +1367,7 @@ class Propal extends CommonObject
 
 		// Clear fields
 		$object->user_author = $user->id;
-		$object->user_validation_id = 0;
+		$object->user_valid = '';
 		$object->date = $now;
 		$object->datep = $now; // deprecated
 		$object->fin_validite = $object->date + ($object->duree_validite * 24 * 3600);
@@ -1411,8 +1410,6 @@ class Propal extends CommonObject
 				$action = '';
 				$reshook = $hookmanager->executeHooks('createFrom', $parameters, $object, $action); // Note that $action and $object may have been modified by some hooks
 				if ($reshook < 0) {
-					$this->errors += $hookmanager->errors;
-					$this->error = $hookmanager->error;
 					$error++;
 				}
 			}
@@ -1553,7 +1550,7 @@ class Propal extends CommonObject
 				$this->extraparams = (array) json_decode($obj->extraparams, true);
 
 				$this->user_author_id = $obj->fk_user_author;
-				$this->user_validation_id = $obj->fk_user_valid;
+				$this->user_valid_id = $obj->fk_user_valid;
 				$this->user_close_id = $obj->fk_user_cloture;
 
 				//Incoterms
@@ -1655,7 +1652,7 @@ class Propal extends CommonObject
 		$sql .= " total_ttc=".(isset($this->total_ttc) ? $this->total_ttc : "null").",";
 		$sql .= " fk_statut=".(isset($this->statut) ? $this->statut : "null").",";
 		$sql .= " fk_user_author=".(isset($this->user_author_id) ? $this->user_author_id : "null").",";
-		$sql .= " fk_user_valid = ".(!empty($this->user_validation_id) ? (int) $this->user_validation_id : "null").",";
+		$sql .= " fk_user_valid=".(isset($this->user_valid) ? $this->user_valid : "null").",";
 		$sql .= " fk_projet=".(isset($this->fk_project) ? $this->fk_project : "null").",";
 		$sql .= " fk_cond_reglement=".(isset($this->cond_reglement_id) ? $this->cond_reglement_id : "null").",";
 		$sql .= " fk_mode_reglement=".(isset($this->mode_reglement_id) ? $this->mode_reglement_id : "null").",";
@@ -1941,7 +1938,7 @@ class Propal extends CommonObject
 			$this->ref = $num;
 			$this->brouillon = 0;
 			$this->statut = self::STATUS_VALIDATED;
-			$this->user_validation_id = $user->id;
+			$this->user_valid_id = $user->id;
 			$this->datev = $now;
 
 			$this->db->commit();
@@ -3198,7 +3195,9 @@ class Propal extends CommonObject
 				$this->user_creation = $cuser;
 
 				if ($obj->fk_user_valid) {
-					$this->user_validation_id = $obj->fk_user_valid;
+					$vuser = new User($this->db);
+					$vuser->fetch($obj->fk_user_valid);
+					$this->user_validation = $vuser;
 				}
 
 				if ($obj->fk_user_signature) {
@@ -4167,40 +4166,36 @@ class PropaleLigne extends CommonObjectLine
 		$error = 0;
 		$this->db->begin();
 
-		if (!$notrigger) {
-			// Call trigger
-			$result = $this->call_trigger('LINEPROPAL_DELETE', $user);
-			if ($result < 0) {
-				$error++;
+		$sql = "DELETE FROM ".MAIN_DB_PREFIX."propaldet WHERE rowid = ".((int) $this->rowid);
+		dol_syslog("PropaleLigne::delete", LOG_DEBUG);
+		if ($this->db->query($sql)) {
+			// Remove extrafields
+			if (!$error) {
+				$this->id = $this->rowid;
+				$result = $this->deleteExtraFields();
+				if ($result < 0) {
+					$error++;
+					dol_syslog(get_class($this)."::delete error -4 ".$this->error, LOG_ERR);
+				}
 			}
-		}
-		// End call triggers
 
-		if (!$error) {
-			$sql = "DELETE FROM " . MAIN_DB_PREFIX . "propaldet WHERE rowid = " . ((int) $this->rowid);
-			dol_syslog("PropaleLigne::delete", LOG_DEBUG);
-			if ($this->db->query($sql)) {
-				// Remove extrafields
-				if (!$error) {
-					$this->id = $this->rowid;
-					$result = $this->deleteExtraFields();
-					if ($result < 0) {
-						$error++;
-						dol_syslog(get_class($this) . "::delete error -4 " . $this->error, LOG_ERR);
-					}
+			if (!$error && !$notrigger) {
+				// Call trigger
+				$result = $this->call_trigger('LINEPROPAL_DELETE', $user);
+				if ($result < 0) {
+					$this->db->rollback();
+					return -1;
 				}
-			} else {
-				$this->error = $this->db->error() . " sql=" . $sql;
-				$error++;
 			}
-		}
+			// End call triggers
 
-		if ($error) {
-			$this->db->rollback();
-			return -1;
-		} else {
 			$this->db->commit();
+
 			return 1;
+		} else {
+			$this->error = $this->db->error()." sql=".$sql;
+			$this->db->rollback();
+			return -1;
 		}
 	}
 
diff --git a/htdocs/comm/propal/class/propalestats.class.php b/htdocs/comm/propal/class/propalestats.class.php
index af660ffd58d..181e5b545f8 100644
--- a/htdocs/comm/propal/class/propalestats.class.php
+++ b/htdocs/comm/propal/class/propalestats.class.php
@@ -93,6 +93,9 @@ class PropaleStats extends Stats
 		}
 		//$this->where.= " AND p.fk_soc = s.rowid AND p.entity = ".$conf->entity;
 		$this->where .= ($this->where ? ' AND ' : '')."p.entity IN (".getEntity('propal').")";
+		if (!$user->rights->societe->client->voir && !$this->socid) {
+			$this->where .= " AND p.fk_soc = sc.fk_soc AND sc.fk_user = ".((int) $user->id);
+		}
 		if ($this->socid) {
 			$this->where .= " AND p.fk_soc = ".((int) $this->socid);
 		}
@@ -127,7 +130,7 @@ class PropaleStats extends Stats
 		$sql = "SELECT date_format(".$this->field_date.",'%m') as dm, COUNT(*) as nb";
 		$sql .= " FROM ".$this->from;
 		if (!$user->rights->societe->client->voir && !$user->socid) {
-			$sql .= " INNER JOIN ".MAIN_DB_PREFIX."societe_commerciaux as sc ON p.fk_soc = sc.fk_soc AND sc.fk_user = ".((int) $user->id);
+			$sql .= ", ".MAIN_DB_PREFIX."societe_commerciaux as sc";
 		}
 		$sql .= $this->join;
 		$sql .= " WHERE ".$this->field_date." BETWEEN '".$this->db->idate(dol_get_first_day($year))."' AND '".$this->db->idate(dol_get_last_day($year))."'";
@@ -152,7 +155,7 @@ class PropaleStats extends Stats
 		$sql = "SELECT date_format(".$this->field_date.",'%Y') as dm, COUNT(*) as nb, SUM(c.".$this->field.")";
 		$sql .= " FROM ".$this->from;
 		if (!$user->rights->societe->client->voir && !$this->socid) {
-			$sql .= "  INNER JOIN ".MAIN_DB_PREFIX."societe_commerciaux as sc ON p.fk_soc = sc.fk_soc AND sc.fk_user = ".((int) $user->id);
+			$sql .= ", ".MAIN_DB_PREFIX."societe_commerciaux as sc";
 		}
 		$sql .= $this->join;
 		$sql .= " WHERE ".$this->where;
@@ -176,7 +179,7 @@ class PropaleStats extends Stats
 		$sql = "SELECT date_format(".$this->field_date.",'%m') as dm, SUM(p.".$this->field.")";
 		$sql .= " FROM ".$this->from;
 		if (!$user->rights->societe->client->voir && !$this->socid) {
-			$sql .= "  INNER JOIN ".MAIN_DB_PREFIX."societe_commerciaux as sc ON p.fk_soc = sc.fk_soc AND sc.fk_user = ".((int) $user->id);
+			$sql .= ", ".MAIN_DB_PREFIX."societe_commerciaux as sc";
 		}
 		$sql .= $this->join;
 		$sql .= " WHERE ".$this->field_date." BETWEEN '".$this->db->idate(dol_get_first_day($year))."' AND '".$this->db->idate(dol_get_last_day($year))."'";
@@ -201,7 +204,7 @@ class PropaleStats extends Stats
 		$sql = "SELECT date_format(".$this->field_date.",'%m') as dm, AVG(p.".$this->field.")";
 		$sql .= " FROM ".$this->from;
 		if (!$user->rights->societe->client->voir && !$this->socid) {
-			$sql .= "  INNER JOIN ".MAIN_DB_PREFIX."societe_commerciaux as sc ON p.fk_soc = sc.fk_soc AND sc.fk_user = ".((int) $user->id);
+			$sql .= ", ".MAIN_DB_PREFIX."societe_commerciaux as sc";
 		}
 		$sql .= $this->join;
 		$sql .= " WHERE ".$this->field_date." BETWEEN '".$this->db->idate(dol_get_first_day($year))."' AND '".$this->db->idate(dol_get_last_day($year))."'";
@@ -224,7 +227,7 @@ class PropaleStats extends Stats
 		$sql = "SELECT date_format(".$this->field_date.",'%Y') as year, COUNT(*) as nb, SUM(".$this->field.") as total, AVG(".$this->field.") as avg";
 		$sql .= " FROM ".$this->from;
 		if (!$user->rights->societe->client->voir && !$this->socid) {
-			$sql .= "  INNER JOIN ".MAIN_DB_PREFIX."societe_commerciaux as sc ON p.fk_soc = sc.fk_soc AND sc.fk_user = ".((int) $user->id);
+			$sql .= ", ".MAIN_DB_PREFIX."societe_commerciaux as sc";
 		}
 		$sql .= $this->join;
 		$sql .= " WHERE ".$this->where;
@@ -250,7 +253,7 @@ class PropaleStats extends Stats
 		$sql = "SELECT product.ref, COUNT(product.ref) as nb, SUM(tl.".$this->field_line.") as total, AVG(tl.".$this->field_line.") as avg";
 		$sql .= " FROM ".$this->from.", ".$this->from_line.", ".MAIN_DB_PREFIX."product as product";
 		if (!$user->rights->societe->client->voir && !$user->socid) {
-			$sql .= "  INNER JOIN ".MAIN_DB_PREFIX."societe_commerciaux as sc ON p.fk_soc = sc.fk_soc AND sc.fk_user = ".((int) $user->id);
+			$sql .= ", ".MAIN_DB_PREFIX."societe_commerciaux as sc";
 		}
 		$sql .= $this->join;
 		$sql .= " WHERE ".$this->where;
diff --git a/htdocs/comm/propal/list.php b/htdocs/comm/propal/list.php
index 037ce5f26f8..3e467539c88 100644
--- a/htdocs/comm/propal/list.php
+++ b/htdocs/comm/propal/list.php
@@ -349,9 +349,9 @@ if ($action == 'validate' && $permissiontovalidate) {
 		$db->begin();
 		$error = 0;
 		foreach ($toselect as $checked) {
-			if ($tmpproposal->fetch($checked) > 0) {
-				if ($tmpproposal->statut == $tmpproposal::STATUS_DRAFT) {
-					if ($tmpproposal->valid($user) > 0) {
+			if ($tmpproposal->fetch($checked)) {
+				if ($tmpproposal->statut == 0) {
+					if ($tmpproposal->valid($user)) {
 						setEventMessage($tmpproposal->ref." ".$langs->trans('PassedInOpenStatus'), 'mesgs');
 					} else {
 						setEventMessage($langs->trans('CantBeValidated'), 'errors');
@@ -362,7 +362,7 @@ if ($action == 'validate' && $permissiontovalidate) {
 					$error++;
 				}
 			} else {
-				setEventMessages($tmpproposal->error, $tmpproposal->errors, 'errors');
+				dol_print_error($db);
 				$error++;
 			}
 		}
@@ -380,13 +380,13 @@ if ($action == "sign" && $permissiontoclose) {
 		$db->begin();
 		$error = 0;
 		foreach ($toselect as $checked) {
-			if ($tmpproposal->fetch($checked) > 0) {
+			if ($tmpproposal->fetch($checked)) {
 				if ($tmpproposal->statut == $tmpproposal::STATUS_VALIDATED) {
-					$tmpproposal->statut = $tmpproposal::STATUS_SIGNED;
-					if ($tmpproposal->closeProposal($user, $tmpproposal::STATUS_SIGNED) >= 0) {
+					$tmpproposal->statut = $tmpproposal::STATUS_SIGNED;;
+					if ($tmpproposal->closeProposal($user, $tmpproposal::STATUS_SIGNED)) {
 						setEventMessage($tmpproposal->ref." ".$langs->trans('Signed'), 'mesgs');
 					} else {
-						setEventMessages($tmpproposal->error, $tmpproposal->errors, 'errors');
+						dol_print_error($db);
 						$error++;
 					}
 				} else {
@@ -394,7 +394,7 @@ if ($action == "sign" && $permissiontoclose) {
 					$error++;
 				}
 			} else {
-				setEventMessages($tmpproposal->error, $tmpproposal->errors, 'errors');
+				dol_print_error($db);
 				$error++;
 			}
 		}
@@ -405,28 +405,27 @@ if ($action == "sign" && $permissiontoclose) {
 		}
 	}
 }
-
 if ($action == "nosign" && $permissiontoclose) {
 	if (GETPOST('confirm') == 'yes') {
 		$tmpproposal = new Propal($db);
 		$db->begin();
 		$error = 0;
 		foreach ($toselect as $checked) {
-			if ($tmpproposal->fetch($checked) > 0) {
+			if ($tmpproposal->fetch($checked)) {
 				if ($tmpproposal->statut == $tmpproposal::STATUS_VALIDATED) {
 					$tmpproposal->statut = $tmpproposal::STATUS_NOTSIGNED;
-					if ($tmpproposal->closeProposal($user, $tmpproposal::STATUS_NOTSIGNED) > 0) {
+					if ($tmpproposal->closeProposal($user, $tmpproposal::STATUS_NOTSIGNED)) {
 						setEventMessage($tmpproposal->ref." ".$langs->trans('NoSigned'), 'mesgs');
 					} else {
-						setEventMessages($tmpproposal->error, $tmpproposal->errors, 'errors');
+						dol_print_error($db);
 						$error++;
 					}
 				} else {
-					setEventMessage($tmpproposal->ref." ".$langs->trans('CantBeNoSign'), 'errors');
+					setEventMessage($tmpproposal->ref." ".$langs->trans('CantBeClosed'), 'errors');
 					$error++;
 				}
 			} else {
-				setEventMessages($tmpproposal->error, $tmpproposal->errors, 'errors');
+				dol_print_error($db);
 				$error++;
 			}
 		}
@@ -1915,7 +1914,7 @@ if ($resql) {
 		// Note public
 		if (!empty($arrayfields['p.note_public']['checked'])) {
 			print '<td class="center">';
-			print dol_string_nohtmltag($obj->note_public);
+			print dol_escape_htmltag($obj->note_public);
 			print '</td>';
 			if (!$i) {
 				$totalarray['nbfield']++;
@@ -1924,7 +1923,7 @@ if ($resql) {
 		// Note private
 		if (!empty($arrayfields['p.note_private']['checked'])) {
 			print '<td class="center">';
-			print dol_string_nohtmltag($obj->note_private);
+			print dol_escape_htmltag($obj->note_private);
 			print '</td>';
 			if (!$i) {
 				$totalarray['nbfield']++;
diff --git a/htdocs/comm/remx.php b/htdocs/comm/remx.php
index 1e3f56c30db..bbf09941f0a 100644
--- a/htdocs/comm/remx.php
+++ b/htdocs/comm/remx.php
@@ -62,24 +62,27 @@ if (GETPOST('cancel', 'alpha') && !empty($backtopage)) {
 	 exit;
 }
 
-if ($action == 'confirm_split' && GETPOST("confirm", "alpha") == 'yes' && ($user->rights->societe->creer || $user->rights->facture->creer)) {
+if ($action == 'confirm_split' && GETPOST("confirm", "alpha") == 'yes' && $user->rights->societe->creer) {
 	//if ($user->rights->societe->creer)
 	//if ($user->rights->facture->creer)
 
 	$amount_ttc_1 = GETPOST('amount_ttc_1', 'alpha');
 	$amount_ttc_1 = price2num($amount_ttc_1);
-	$amount_ttc_2 = GETPOST('amount_ttc_2', 'alpha');
-	$amount_ttc_2 = price2num($amount_ttc_2);
+//	$amount_ttc_2 = GETPOST('amount_ttc_2', 'alpha');
+//	$amount_ttc_2 = price2num($amount_ttc_2);
 
 	$error = 0;
 	$remid = (GETPOST("remid", 'int') ? GETPOST("remid", 'int') : 0);
 	$discount = new DiscountAbsolute($db);
 	$res = $discount->fetch($remid);
+
+	$amount_ttc_2 = price2num($discount->amount_ttc - $amount_ttc_1);
+
 	if (!($res > 0)) {
 		$error++;
 		setEventMessages($langs->trans("ErrorFailedToLoadDiscount"), null, 'errors');
 	}
-	if (!$error && price2num($amount_ttc_1 + $amount_ttc_2) != $discount->amount_ttc) {
+	if (!$error && (price2num($amount_ttc_1 + $amount_ttc_2) != $discount->amount_ttc || $amount_ttc_1 <= 0 || $amount_ttc_2 <= 0)) {
 		$error++;
 		setEventMessages($langs->trans("TotalOfTwoDiscountMustEqualsOriginal"), null, 'errors');
 	}
@@ -129,7 +132,7 @@ if ($action == 'confirm_split' && GETPOST("confirm", "alpha") == 'yes' && ($user
 		$newdiscount1->amount_tva = price2num($newdiscount1->amount_ttc - $newdiscount1->amount_ht);
 		$newdiscount2->amount_tva = price2num($newdiscount2->amount_ttc - $newdiscount2->amount_ht);
 
-		$newdiscount1->multicurrency_amount_ttc = $amount_ttc_1 * ($discount->multicurrency_amount_ttc / $discount->amount_ttc);
+		$newdiscount1->multicurrency_amount_ttc = price2num($amount_ttc_1 * ($discount->multicurrency_amount_ttc / $discount->amount_ttc));
 		$newdiscount2->multicurrency_amount_ttc = price2num($discount->multicurrency_amount_ttc - $newdiscount1->multicurrency_amount_ttc);
 		$newdiscount1->multicurrency_amount_ht = price2num($newdiscount1->multicurrency_amount_ttc / (1 + $newdiscount1->tva_tx / 100), 'MT');
 		$newdiscount2->multicurrency_amount_ht = price2num($newdiscount2->multicurrency_amount_ttc / (1 + $newdiscount2->tva_tx / 100), 'MT');
@@ -153,7 +156,102 @@ if ($action == 'confirm_split' && GETPOST("confirm", "alpha") == 'yes' && ($user
 	}
 }
 
-if ($action == 'setremise' && ($user->rights->societe->creer || $user->rights->facture->creer)) {
+if ($action == 'confirm_split_multicurrency' && GETPOST("confirm", "alpha") == 'yes' && $user->rights->societe->creer) {
+	//if ($user->rights->societe->creer)
+	//if ($user->rights->facture->creer)
+
+	$multicurrency_amount_ttc_1 = GETPOST('multicurrency_amount_ttc_1', 'alpha');
+	$multicurrency_amount_ttc_1 = price2num($multicurrency_amount_ttc_1);
+//	$amount_ttc_2 = GETPOST('amount_ttc_2', 'alpha');
+//	$amount_ttc_2 = price2num($amount_ttc_2);
+
+	$error = 0;
+	$remid = (GETPOST("remid", 'int') ? GETPOST("remid", 'int') : 0);
+	$discount = new DiscountAbsolute($db);
+	$res = $discount->fetch($remid);
+
+	$multicurrency_amount_ttc_2 = price2num($discount->multicurrency_amount_ttc - $multicurrency_amount_ttc_1);
+
+	if (!($res > 0)) {
+		$error++;
+		setEventMessages($langs->trans("ErrorFailedToLoadDiscount"), null, 'errors');
+	}
+	if (!$error && (price2num($multicurrency_amount_ttc_1 + $multicurrency_amount_ttc_2) != $discount->multicurrency_amount_ttc || $multicurrency_amount_ttc_1 <= 0 || $multicurrency_amount_ttc_2 <= 0)) {
+		$error++;
+		setEventMessages($langs->trans("TotalOfTwoDiscountMustEqualsOriginal"), null, 'errors');
+	}
+	if (!$error && $discount->fk_facture_line) {
+		$error++;
+		setEventMessages($langs->trans("ErrorCantSplitAUsedDiscount"), null, 'errors');
+	}
+	if (!$error) {
+		$newdiscount1 = new DiscountAbsolute($db);
+		$newdiscount2 = new DiscountAbsolute($db);
+		$newdiscount1->fk_facture_source = $discount->fk_facture_source;
+		$newdiscount2->fk_facture_source = $discount->fk_facture_source;
+		$newdiscount1->fk_facture = $discount->fk_facture;
+		$newdiscount2->fk_facture = $discount->fk_facture;
+		$newdiscount1->fk_facture_line = $discount->fk_facture_line;
+		$newdiscount2->fk_facture_line = $discount->fk_facture_line;
+		$newdiscount1->fk_invoice_supplier_source = $discount->fk_invoice_supplier_source;
+		$newdiscount2->fk_invoice_supplier_source = $discount->fk_invoice_supplier_source;
+		$newdiscount1->fk_invoice_supplier = $discount->fk_invoice_supplier;
+		$newdiscount2->fk_invoice_supplier = $discount->fk_invoice_supplier;
+		$newdiscount1->fk_invoice_supplier_line = $discount->fk_invoice_supplier_line;
+		$newdiscount2->fk_invoice_supplier_line = $discount->fk_invoice_supplier_line;
+		if ($discount->description == '(CREDIT_NOTE)' || $discount->description == '(DEPOSIT)') {
+			$newdiscount1->description = $discount->description;
+			$newdiscount2->description = $discount->description;
+		} else {
+			$newdiscount1->description = $discount->description.' (1)';
+			$newdiscount2->description = $discount->description.' (2)';
+		}
+
+		$newdiscount1->fk_user = $discount->fk_user;
+		$newdiscount2->fk_user = $discount->fk_user;
+		$newdiscount1->fk_soc = $discount->fk_soc;
+		$newdiscount2->fk_soc = $discount->fk_soc;
+		$newdiscount1->discount_type = $discount->discount_type;
+		$newdiscount2->discount_type = $discount->discount_type;
+		$newdiscount1->datec = $discount->datec;
+		$newdiscount2->datec = $discount->datec;
+		$newdiscount1->tva_tx = $discount->tva_tx;
+		$newdiscount2->tva_tx = $discount->tva_tx;
+		$newdiscount1->vat_src_code = $discount->vat_src_code;
+		$newdiscount2->vat_src_code = $discount->vat_src_code;
+
+		$newdiscount1->multicurrency_amount_ttc = $multicurrency_amount_ttc_1;
+		$newdiscount1->amount_ttc = price2num($multicurrency_amount_ttc_1 / ($discount->multicurrency_amount_ttc / $discount->amount_ttc));
+		$newdiscount2->amount_ttc = price2num($discount->amount_ttc - $newdiscount1->amount_ttc);
+		$newdiscount1->amount_ht = price2num($newdiscount1->amount_ttc / (1 + $newdiscount1->tva_tx / 100), 'MT');
+		$newdiscount2->amount_ht = price2num($newdiscount2->amount_ttc / (1 + $newdiscount2->tva_tx / 100), 'MT');
+		$newdiscount1->amount_tva = price2num($newdiscount1->amount_ttc - $newdiscount1->amount_ht);
+		$newdiscount2->amount_tva = price2num($newdiscount2->amount_ttc - $newdiscount2->amount_ht);
+
+		$newdiscount2->multicurrency_amount_ttc = price2num($discount->multicurrency_amount_ttc - $newdiscount1->multicurrency_amount_ttc);
+		$newdiscount1->multicurrency_amount_ht = price2num($newdiscount1->multicurrency_amount_ttc / (1 + $newdiscount1->tva_tx / 100), 'MT');
+		$newdiscount2->multicurrency_amount_ht = price2num($newdiscount2->multicurrency_amount_ttc / (1 + $newdiscount2->tva_tx / 100), 'MT');
+		$newdiscount1->multicurrency_amount_tva = price2num($newdiscount1->multicurrency_amount_ttc - $newdiscount1->multicurrency_amount_ht);
+		$newdiscount2->multicurrency_amount_tva = price2num($newdiscount2->multicurrency_amount_ttc - $newdiscount2->multicurrency_amount_ht);
+
+		$db->begin();
+		$discount->fk_facture_source = 0; // This is to delete only the require record (that we will recreate with two records) and not all family with same fk_facture_source
+		// This is to delete only the require record (that we will recreate with two records) and not all family with same fk_invoice_supplier_source
+		$discount->fk_invoice_supplier_source = 0;
+		$res = $discount->delete($user);
+		$newid1 = $newdiscount1->create($user);
+		$newid2 = $newdiscount2->create($user);
+		if ($res > 0 && $newid1 > 0 && $newid2 > 0) {
+			$db->commit();
+			header("Location: ".$_SERVER["PHP_SELF"].'?id='.$id.($backtopage ? '&backtopage='.urlencode($backtopage) : '')); // To avoid pb whith back
+			exit;
+		} else {
+			$db->rollback();
+		}
+	}
+}
+
+if ($action == 'setremise' && $user->rights->societe->creer) {
 	//if ($user->rights->societe->creer)
 	//if ($user->rights->facture->creer)
 
@@ -192,7 +290,7 @@ if ($action == 'setremise' && ($user->rights->societe->creer || $user->rights->f
 	}
 }
 
-if (GETPOST('action', 'aZ09') == 'confirm_remove' && GETPOST("confirm") == 'yes' && ($user->rights->societe->creer || $user->rights->facture->creer)) {
+if (GETPOST('action', 'aZ09') == 'confirm_remove' && GETPOST("confirm") == 'yes' && $user->rights->societe->creer) {
 	//if ($user->rights->societe->creer)
 	//if ($user->rights->facture->creer)
 
@@ -429,14 +527,19 @@ if ($socid > 0) {
 			}
 			print '<td class="right">'.$langs->trans("VATRate").'</td>';
 			print '<td class="right">'.$langs->trans("AmountTTC").'</td>';
+			print '<td>&nbsp;</td>';
 			if (!empty($conf->multicurrency->enabled)) {
 				print '<td class="right">'.$langs->trans("MulticurrencyAmountTTC").'</td>';
+				print '<td>&nbsp;</td>';
 			}
 			print '<td width="100" class="center">'.$langs->trans("DiscountOfferedBy").'</td>';
 			print '<td width="50">&nbsp;</td>';
 			print '</tr>';
 
 			$showconfirminfo = array();
+			if (!empty($conf->multicurrency->enabled)) {
+				$showconfirminfo_multicurrency = array();
+			}
 
 			$i = 0;
 			$num = $db->num_rows($resql);
@@ -479,15 +582,28 @@ if ($socid > 0) {
 					}
 					print '<td class="right">'.vatrate($obj->tva_tx.($obj->vat_src_code ? ' ('.$obj->vat_src_code.')' : ''), true).'</td>';
 					print '<td class="right">'.price($obj->amount_ttc).'</td>';
+					if ($user->rights->societe->creer || $user->rights->facture->creer) {
+						print '<td class="center nowrap">';
+						print '<a class="reposition" href="'.$_SERVER["PHP_SELF"].'?id='.$object->id.'&action=split&token='.newToken().'&remid='.$obj->rowid.($backtopage ? '&backtopage='.urlencode($backtopage) : '').'">'.img_split($langs->trans("SplitDiscount")).'</a>';
+						print '</td>';
+					} else {
+						print '<td>&nbsp;</td>';
+					}
 					if (!empty($conf->multicurrency->enabled)) {
 						print '<td class="right">'.price($obj->multicurrency_amount_ttc).'</td>';
+						if ($user->rights->societe->creer || $user->rights->facture->creer) {
+							print '<td class="center nowrap">';
+							print '<a class="reposition" href="'.$_SERVER["PHP_SELF"].'?id='.$object->id.'&action=split_multicurrency&token='.newToken().'&remid='.$obj->rowid.($backtopage ? '&backtopage='.urlencode($backtopage) : '').'">'.img_split($langs->trans("SplitDiscount")).'</a>';
+							print '</td>';
+						} else {
+							print '<td>&nbsp;</td>';
+						}
 					}
 					print '<td class="center">';
 					print '<a href="'.DOL_URL_ROOT.'/user/card.php?id='.$obj->user_id.'">'.img_object($langs->trans("ShowUser"), 'user').' '.$obj->login.'</a>';
 					print '</td>';
 					if ($user->rights->societe->creer || $user->rights->facture->creer) {
 						print '<td class="center nowrap">';
-						print '<a class="reposition" href="'.$_SERVER["PHP_SELF"].'?id='.$object->id.'&action=split&token='.newToken().'&remid='.$obj->rowid.($backtopage ? '&backtopage='.urlencode($backtopage) : '').'">'.img_split($langs->trans("SplitDiscount")).'</a>';
 						print '<a class="reposition marginleftonly" href="'.$_SERVER["PHP_SELF"].'?id='.$object->id.'&action=remove&token='.newToken().'&remid='.$obj->rowid.($backtopage ? '&backtopage='.urlencode($backtopage) : '').'">'.img_delete($langs->trans("RemoveDiscount")).'</a>';
 						print '</td>';
 					} else {
@@ -499,10 +615,14 @@ if ($socid > 0) {
 						$showconfirminfo['rowid'] = $obj->rowid;
 						$showconfirminfo['amount_ttc'] = $obj->amount_ttc;
 					}
+					elseif ($_GET["action"] == 'split_multicurrency' && GETPOST('remid') == $obj->rowid) {
+						$showconfirminfo_multicurrency['rowid'] = $obj->rowid;
+						$showconfirminfo_multicurrency['multicurrency_amount_ttc'] = $obj->multicurrency_amount_ttc;
+					}
 					$i++;
 				}
 			} else {
-				$colspan = 8;
+				$colspan = 9;
 				if (!empty($conf->multicurrency->enabled)) {
 					$colspan += 2;
 				}
@@ -514,14 +634,24 @@ if ($socid > 0) {
 
 			if (count($showconfirminfo)) {
 				$amount1 = price2num($showconfirminfo['amount_ttc'] / 2, 'MT');
-				$amount2 = ($showconfirminfo['amount_ttc'] - $amount1);
+			//	$amount2 = ($showconfirminfo['amount_ttc'] - $amount1);
 				$formquestion = array(
 					'text' => $langs->trans('TypeAmountOfEachNewDiscount'),
-					array('type' => 'text', 'name' => 'amount_ttc_1', 'label' => $langs->trans("AmountTTC").' 1', 'value' => $amount1, 'size' => '5'),
-					array('type' => 'text', 'name' => 'amount_ttc_2', 'label' => $langs->trans("AmountTTC").' 2', 'value' => $amount2, 'size' => '5')
+					array('type' => 'text', 'id'=> 'amount_ttc_1', 'name' => 'amount_ttc_1', 'label' => $langs->trans("AmountTTC").' 1', 'value' => $amount1, 'size' => '5')
+			//		array('type' => 'text', 'id'=> 'amount_ttc_2', 'name' => 'amount_ttc_2', 'label' => $langs->trans("AmountTTC").' 2', 'value' => $amount2, 'size' => '5')
 				);
 				$langs->load("dict");
-				print $form->formconfirm($_SERVER["PHP_SELF"].'?id='.$object->id.'&remid='.$showconfirminfo['rowid'].($backtopage ? '&backtopage='.urlencode($backtopage) : ''), $langs->trans('SplitDiscount'), $langs->trans('ConfirmSplitDiscount', price($showconfirminfo['amount_ttc']), $langs->transnoentities("Currency".$conf->currency)), 'confirm_split', $formquestion, '', 0);
+				print $form->formconfirm($_SERVER["PHP_SELF"].'?id='.$object->id.'&remid='.$showconfirminfo['rowid'].($backtopage ? '&backtopage='.urlencode($backtopage) : ''), $langs->trans('SplitDiscount'), $langs->trans('ConfirmSplitDiscount', price($showconfirminfo['amount_ttc']), $langs->transnoentities("Currency".$conf->currency)), 'confirm_split', $formquestion, '', 1);
+			} elseif (count($showconfirminfo_multicurrency)) {
+				$multicurrency_amount1 = price2num($showconfirminfo_multicurrency['multicurrency_amount_ttc'] / 2, 'MT');
+			//	$multicurrency_amount2 = ($showconfirminfo['multicurrency_amount_ttc'] - $amount1);
+				$formquestion = array(
+					'text' => $langs->trans('TypeAmountOfEachNewDiscount'),
+					array('type' => 'text', 'id'=> 'multicurrency_amount_ttc_1', 'name' => 'multicurrency_amount_ttc_1', 'label' => $langs->trans("MulticurrencyAmountTTC").' 1', 'value' => $multicurrency_amount1, 'size' => '5')
+			//		array('type' => 'text', 'id'=> 'multicurrency_amount_ttc_2', 'name' => 'multicurrency_amount_ttc_2', 'label' => $langs->trans("MulticurrencyAmountTTC").' 2', 'value' => $multicurrency_amount2, 'size' => '5')
+				);
+				$langs->load("dict");
+				print $form->formconfirm($_SERVER["PHP_SELF"].'?id='.$object->id.'&remid='.$showconfirminfo_multicurrency['rowid'].($backtopage ? '&backtopage='.urlencode($backtopage) : ''), $langs->trans('SplitDiscount'), $langs->trans('ConfirmSplitDiscount', price($showconfirminfo_multicurrency['multicurrency_amount_ttc']), $langs->trans("MulticurrencyAmountTTC")), 'confirm_split_multicurrency', $formquestion, '', 1);
 			}
 		} else {
 			dol_print_error($db);
@@ -568,8 +698,10 @@ if ($socid > 0) {
 			}
 			print '<td class="right">'.$langs->trans("VATRate").'</td>';
 			print '<td class="right">'.$langs->trans("AmountTTC").'</td>';
+			print '<td>&nbsp;</td>';
 			if (!empty($conf->multicurrency->enabled)) {
 				print '<td class="right">'.$langs->trans("MulticurrencyAmountTTC").'</td>';
+				print '<td>&nbsp;</td>';
 			}
 			print '<td width="100" class="center">'.$langs->trans("DiscountOfferedBy").'</td>';
 			print '<td width="50">&nbsp;</td>';
@@ -618,15 +750,28 @@ if ($socid > 0) {
 					}
 					print '<td class="right">'.vatrate($obj->tva_tx.($obj->vat_src_code ? ' ('.$obj->vat_src_code.')' : ''), true).'</td>';
 					print '<td class="right">'.price($obj->amount_ttc).'</td>';
+					if ($user->rights->societe->creer || $user->rights->facture->creer) {
+						print '<td class="center nowrap">';
+						print '<a class="reposition" href="'.$_SERVER["PHP_SELF"].'?id='.$object->id.'&action=split&token='.newToken().'&remid='.$obj->rowid.($backtopage ? '&backtopage='.urlencode($backtopage) : '').'">'.img_split($langs->trans("SplitDiscount")).'</a>';
+						print '</td>';
+					} else {
+						print '<td>&nbsp;</td>';
+					}
 					if (!empty($conf->multicurrency->enabled)) {
 						print '<td class="right">'.price($obj->multicurrency_amount_ttc).'</td>';
+						if ($user->rights->societe->creer || $user->rights->facture->creer) {
+							print '<td class="center nowrap">';
+							print '<a class="reposition" href="'.$_SERVER["PHP_SELF"].'?id='.$object->id.'&action=split_multicurrency&token='.newToken().'&remid='.$obj->rowid.($backtopage ? '&backtopage='.urlencode($backtopage) : '').'">'.img_split($langs->trans("SplitDiscount")).'</a>';
+							print '</td>';
+						} else {
+							print '<td>&nbsp;</td>';
+						}
 					}
 					print '<td class="center">';
 					print '<a href="'.DOL_URL_ROOT.'/user/card.php?id='.$obj->user_id.'">'.img_object($langs->trans("ShowUser"), 'user').' '.$obj->login.'</a>';
 					print '</td>';
 					if ($user->rights->societe->creer || $user->rights->facture->creer) {
 						print '<td class="center nowrap">';
-						print '<a class="reposition" href="'.$_SERVER["PHP_SELF"].'?id='.$object->id.'&action=split&token='.newToken().'&remid='.$obj->rowid.($backtopage ? '&backtopage='.urlencode($backtopage) : '').'">'.img_split($langs->trans("SplitDiscount")).'</a>';
 						print '<a class="reposition marginleftonly" href="'.$_SERVER["PHP_SELF"].'?id='.$object->id.'&action=remove&token='.newToken().'&remid='.$obj->rowid.($backtopage ? '&backtopage='.urlencode($backtopage) : '').'">'.img_delete($langs->trans("RemoveDiscount")).'</a>';
 						print '</td>';
 					} else {
@@ -637,11 +782,14 @@ if ($socid > 0) {
 					if ($_GET["action"] == 'split' && GETPOST('remid') == $obj->rowid) {
 						$showconfirminfo['rowid'] = $obj->rowid;
 						$showconfirminfo['amount_ttc'] = $obj->amount_ttc;
+					} elseif ($_GET["action"] == 'split_multicurrency' && GETPOST('remid') == $obj->rowid) {
+						$showconfirminfo_multicurrency['rowid'] = $obj->rowid;
+						$showconfirminfo_multicurrency['multicurrency_amount_ttc'] = $obj->multicurrency_amount_ttc;
 					}
 					$i++;
 				}
 			} else {
-				$colspan = 8;
+				$colspan = 9;
 				if (!empty($conf->multicurrency->enabled)) {
 					$colspan += 2;
 				}
@@ -653,14 +801,24 @@ if ($socid > 0) {
 
 			if (count($showconfirminfo)) {
 				$amount1 = price2num($showconfirminfo['amount_ttc'] / 2, 'MT');
-				$amount2 = ($showconfirminfo['amount_ttc'] - $amount1);
+			//	$amount2 = ($showconfirminfo['amount_ttc'] - $amount1);
+				$formquestion = array(
+					'text' => $langs->trans('TypeAmountOfEachNewDiscount'),
+					array('type' => 'text', 'name' => 'amount_ttc_1', 'label' => $langs->trans("AmountTTC").' 1', 'value' => $amount1, 'size' => '5')
+			//		array('type' => 'text', 'name' => 'amount_ttc_2', 'label' => $langs->trans("AmountTTC").' 2', 'value' => $amount2, 'size' => '5')
+				);
+				$langs->load("dict");
+				print $form->formconfirm($_SERVER["PHP_SELF"].'?id='.$object->id.'&remid='.$showconfirminfo['rowid'].($backtopage ? '&backtopage='.urlencode($backtopage) : ''), $langs->trans('SplitDiscount'), $langs->trans('ConfirmSplitDiscount', price($showconfirminfo['amount_ttc']), $langs->transnoentities("Currency".$conf->currency)), 'confirm_split', $formquestion, 0, 1);
+			} elseif (count($showconfirminfo_multicurrency)) {
+				$multicurrency_amount1 = price2num($showconfirminfo_multicurrency['multicurrency_amount_ttc'] / 2, 'MT');
+			//	$multicurrency_amount2 = ($showconfirminfo['multicurrency_amount_ttc'] - $amount1);
 				$formquestion = array(
 					'text' => $langs->trans('TypeAmountOfEachNewDiscount'),
-					array('type' => 'text', 'name' => 'amount_ttc_1', 'label' => $langs->trans("AmountTTC").' 1', 'value' => $amount1, 'size' => '5'),
-					array('type' => 'text', 'name' => 'amount_ttc_2', 'label' => $langs->trans("AmountTTC").' 2', 'value' => $amount2, 'size' => '5')
+					array('type' => 'text', 'id'=> 'multicurrency_amount_ttc_1', 'name' => 'multicurrency_amount_ttc_1', 'label' => $langs->trans("MulticurrencyAmountTTC").' 1', 'value' => $multicurrency_amount1, 'size' => '5')
+			//		array('type' => 'text', 'id'=> 'multicurrency_amount_ttc_2', 'name' => 'multicurrency_amount_ttc_2', 'label' => $langs->trans("MulticurrencyAmountTTC").' 2', 'value' => $multicurrency_amount2, 'size' => '5')
 				);
 				$langs->load("dict");
-				print $form->formconfirm($_SERVER["PHP_SELF"].'?id='.$object->id.'&remid='.$showconfirminfo['rowid'].($backtopage ? '&backtopage='.urlencode($backtopage) : ''), $langs->trans('SplitDiscount'), $langs->trans('ConfirmSplitDiscount', price($showconfirminfo['amount_ttc']), $langs->transnoentities("Currency".$conf->currency)), 'confirm_split', $formquestion, 0, 0);
+				print $form->formconfirm($_SERVER["PHP_SELF"].'?id='.$object->id.'&remid='.$showconfirminfo_multicurrency['rowid'].($backtopage ? '&backtopage='.urlencode($backtopage) : ''), $langs->trans('SplitDiscount'), $langs->trans('ConfirmSplitDiscount', price($showconfirminfo_multicurrency['multicurrency_amount_ttc']), $langs->trans("MulticurrencyAmountTTC")), 'confirm_split_multicurrency', $formquestion, 0, 1);
 			}
 		} else {
 			dol_print_error($db);
@@ -710,7 +868,8 @@ if ($socid > 0) {
 		// Discount linked to invoices
 		$sql2 = "SELECT rc.rowid, rc.amount_ht, rc.amount_tva, rc.amount_ttc, rc.tva_tx, rc.vat_src_code,";
 		$sql2 .= " rc.multicurrency_amount_ht, rc.multicurrency_amount_tva, rc.multicurrency_amount_ttc,";
-		$sql2 .= " rc.datec as dc, rc.description, rc.fk_facture, rc.fk_facture_source,";
+		$sql2 .= " rc.datec as dc, rc.description, rc.fk_facture_line, rc.fk_facture,";
+		$sql2 .= " rc.fk_facture_source,";
 		$sql2 .= " u.login, u.rowid as user_id,";
 		$sql2 .= " f.rowid as invoiceid, f.ref,";
 		$sql2 .= " fa.ref as invoice_source_ref, fa.type as type";
@@ -870,7 +1029,7 @@ if ($socid > 0) {
 		// Discount linked to invoices
 		$sql2 = "SELECT rc.rowid, rc.amount_ht, rc.amount_tva, rc.amount_ttc, rc.tva_tx, rc.vat_src_code,";
 		$sql2 .= " rc.multicurrency_amount_ht, rc.multicurrency_amount_tva, rc.multicurrency_amount_ttc,";
-		$sql2 .= " rc.datec as dc, rc.description, rc.fk_invoice_supplier,";
+		$sql2 .= " rc.datec as dc, rc.description, rc.fk_invoice_supplier_line, rc.fk_invoice_supplier,";
 		$sql2 .= " rc.fk_invoice_supplier_source,";
 		$sql2 .= " u.login, u.rowid as user_id,";
 		$sql2 .= " f.rowid as invoiceid, f.ref as ref,";
