diff --git a/htdocs/commande/card.php b/htdocs/commande/card.php
index aef28d0bf14..2b7a73cd360 100644
--- a/htdocs/commande/card.php
+++ b/htdocs/commande/card.php
@@ -445,7 +445,6 @@ if (empty($reshook)) {
 						// Note that $action and $object may be modified by hook
 						$reshook = $hookmanager->executeHooks('createFrom', $parameters, $object, $action);
 						if ($reshook < 0) {
-							setEventMessages($hookmanager->error, $hookmanager->errors, 'errors');
 							$error++;
 						}
 					} else {
@@ -796,12 +795,17 @@ if (empty($reshook)) {
 					}
 				}
 
-				$tmpvat = (float) price2num(preg_replace('/\s*\(.*\)/', '', $tva_tx));
-				$tmpprodvat = (float) price2num(preg_replace('/\s*\(.*\)/', '', $prod->tva_tx));
+				$tmpvat = price2num(preg_replace('/\s*\(.*\)/', '', $tva_tx));
+				$tmpprodvat = price2num(preg_replace('/\s*\(.*\)/', '', $prod->tva_tx));
 
 				// if price ht is forced (ie: calculated by margin rate and cost price). TODO Why this ?
 				if (!empty($price_ht) || $price_ht === '0') {
-					$pu_ht = price2num($price_ht, 'MU');
+					if($price_ht === '0' && $price_ht_devise != '') { //added test to record multicurrency price while the default price is 0.
+						$pu_ht_devise = price2num($price_ht_devise, 'MU');
+						$pu_ht = price2num(GETPOST('price_ht', 'MU'));
+					} else {
+						$pu_ht = price2num($price_ht, 'MU');
+					}
 					$pu_ttc = price2num($pu_ht * (1 + ($tmpvat / 100)), 'MU');
 				} elseif ($tmpvat != $tmpprodvat) {
 					// On reevalue prix selon taux tva car taux tva transaction peut etre different
@@ -1579,7 +1583,7 @@ if ($action == 'create' && $usercancreate) {
 		print '</td>';
 	} else {
 		print '<td>';
-		print img_picto('', 'company').$form->select_company('', 'socid', '((s.client = 1 OR s.client = 2 OR s.client = 3) AND s.status=1)', 'SelectThirdParty', 0, 0, null, 0, 'minwidth175 maxwidth500 widthcentpercentminusxx');
+		print img_picto('', 'company').$form->select_company('', 'socid', '(s.client = 1 OR s.client = 2 OR s.client = 3)', 'SelectThirdParty', 0, 0, null, 0, 'minwidth175 maxwidth500 widthcentpercentminusxx');
 		// reload page to retrieve customer informations
 		if (empty($conf->global->RELOAD_PAGE_ON_CUSTOMER_CHANGE_DISABLED)) {
 			print '<script type="text/javascript">
@@ -1878,12 +1882,8 @@ if ($action == 'create' && $usercancreate) {
 		if ($action == 'validate') {
 			// We check that object has a temporary ref
 			$ref = substr($object->ref, 1, 4);
-			if ($ref == 'PROV' || $ref == '') {
+			if ($ref == 'PROV') {
 				$numref = $object->getNextNumRef($soc);
-				if (empty($numref)) {
-					$error++;
-					setEventMessages($object->error, $object->errors, 'errors');
-				}
 			} else {
 				$numref = $object->ref;
 			}
@@ -1920,9 +1920,7 @@ if ($action == 'create' && $usercancreate) {
 				);
 			}
 
-			if (!$error) {
-				$formconfirm = $form->formconfirm($_SERVER["PHP_SELF"].'?id='.$object->id, $langs->trans('ValidateOrder'), $text, 'confirm_validate', $formquestion, 0, 1, 220);
-			}
+			$formconfirm = $form->formconfirm($_SERVER["PHP_SELF"].'?id='.$object->id, $langs->trans('ValidateOrder'), $text, 'confirm_validate', $formquestion, 0, 1, 220);
 		}
 
 		// Confirm back to draft status
@@ -2526,7 +2524,7 @@ if ($action == 'create' && $usercancreate) {
 				}
 
 				// Create intervention
-				if (!empty($conf->ficheinter->enabled)) {
+				if ($conf->ficheinter->enabled) {
 					$langs->load("interventions");
 
 					if ($object->statut > Commande::STATUS_DRAFT && $object->statut < Commande::STATUS_CLOSED && $object->getNbOfServicesLines() > 0) {
diff --git a/htdocs/commande/class/api_orders.class.php b/htdocs/commande/class/api_orders.class.php
index b550139cd94..a90afdba4f1 100644
--- a/htdocs/commande/class/api_orders.class.php
+++ b/htdocs/commande/class/api_orders.class.php
@@ -228,7 +228,9 @@ class Orders extends DolibarrApi
 				$commande_static = new Commande($this->db);
 				if ($commande_static->fetch($obj->rowid)) {
 					// Add external contacts ids
-					$commande_static->contacts_ids = $commande_static->liste_contact(-1, 'external', 1);
+					$commande_static->contacts_ids_external = $commande_static->liste_contact(-1, 'external', 0);
+					$commande_static->contacts_ids_internal = $commande_static->liste_contact(-1, 'internal', 0);
+					$commande_static->fetchObjectLinked();
 					$obj_ret[] = $this->_cleanObjectDatas($commande_static);
 				}
 				$i++;
diff --git a/htdocs/commande/class/commande.class.php b/htdocs/commande/class/commande.class.php
index eb0b6378d79..5266c147ebc 100644
--- a/htdocs/commande/class/commande.class.php
+++ b/htdocs/commande/class/commande.class.php
@@ -314,8 +314,8 @@ class Commande extends CommonOrder
 		'localtax2' =>array('type'=>'double(24,8)', 'label'=>'LocalTax2', 'enabled'=>1, 'visible'=>-1, 'position'=>135, 'isameasure'=>1),
 		'total_ht' =>array('type'=>'double(24,8)', 'label'=>'TotalHT', 'enabled'=>1, 'visible'=>-1, 'position'=>140, 'isameasure'=>1),
 		'total_ttc' =>array('type'=>'double(24,8)', 'label'=>'TotalTTC', 'enabled'=>1, 'visible'=>-1, 'position'=>145, 'isameasure'=>1),
-		'note_private' =>array('type'=>'text', 'label'=>'NotePrivate', 'enabled'=>1, 'visible'=>0, 'position'=>150),
-		'note_public' =>array('type'=>'text', 'label'=>'NotePublic', 'enabled'=>1, 'visible'=>0, 'position'=>155),
+		'note_private' =>array('type'=>'text', 'label'=>'NotePublic', 'enabled'=>1, 'visible'=>0, 'position'=>150),
+		'note_public' =>array('type'=>'text', 'label'=>'NotePrivate', 'enabled'=>1, 'visible'=>0, 'position'=>155),
 		'model_pdf' =>array('type'=>'varchar(255)', 'label'=>'PDFTemplate', 'enabled'=>1, 'visible'=>0, 'position'=>160),
 		//'facture' =>array('type'=>'tinyint(4)', 'label'=>'ParentInvoice', 'enabled'=>1, 'visible'=>-1, 'position'=>165),
 		'fk_account' =>array('type'=>'integer', 'label'=>'BankAccount', 'enabled'=>1, 'visible'=>-1, 'position'=>170),
@@ -1156,7 +1156,7 @@ class Commande extends CommonOrder
 				}
 			}
 		} else {
-			$this->error = $this->db->lasterror();
+			dol_print_error($this->db);
 			$this->db->rollback();
 			return -1;
 		}
@@ -1252,8 +1252,6 @@ class Commande extends CommonOrder
 				$action = '';
 				$reshook = $hookmanager->executeHooks('createFrom', $parameters, $this, $action); // Note that $action and $object may have been modified by some hooks
 				if ($reshook < 0) {
-					$this->errors += $hookmanager->errors;
-					$this->error = $hookmanager->error;
 					$error++;
 				}
 			}
@@ -1283,7 +1281,6 @@ class Commande extends CommonOrder
 	{
 		global $conf, $hookmanager;
 
-		dol_include_once('/multicurrency/class/multicurrency.class.php');
 		dol_include_once('/core/class/extrafields.class.php');
 
 		$error = 0;
@@ -1325,9 +1322,6 @@ class Commande extends CommonOrder
 			$line->marge_tx			= $marginInfos[1];
 			$line->marque_tx		= $marginInfos[2];
 
-			$line->origin           = $object->element;
-			$line->origin_id        = $object->lines[$i]->id;
-
 			// get extrafields from original line
 			$object->lines[$i]->fetch_optionals();
 			foreach ($object->lines[$i]->array_options as $options_key => $value) {
@@ -1361,29 +1355,6 @@ class Commande extends CommonOrder
 		$this->origin = $object->element;
 		$this->origin_id = $object->id;
 
-				// Multicurrency (test on $this->multicurrency_tx because we should take the default rate only if not using origin rate)
-		if (!empty($conf->multicurrency->enabled)) {
-			if (!empty($object->multicurrency_code)) {
-				$this->multicurrency_code = $object->multicurrency_code;
-			}
-			if (!empty($conf->global->MULTICURRENCY_USE_ORIGIN_TX) && !empty($object->multicurrency_tx)) {
-				$this->multicurrency_tx = $object->multicurrency_tx;
-			}
-
-			if (!empty($this->multicurrency_code) && empty($this->multicurrency_tx)) {
-					$tmparray = MultiCurrency::getIdAndTxFromCode($this->db, $this->multicurrency_code, $this->date_commande);
-					$this->fk_multicurrency = $tmparray[0];
-					$this->multicurrency_tx = $tmparray[1];
-			} else {
-					$this->fk_multicurrency = MultiCurrency::getIdFromCode($this->db, $this->multicurrency_code);
-			}
-			if (empty($this->fk_multicurrency)) {
-					$this->multicurrency_code = $conf->currency;
-					$this->fk_multicurrency = 0;
-					$this->multicurrency_tx = 1;
-			}
-		}
-
 		// get extrafields from original line
 		$object->fetch_optionals();
 
@@ -1411,8 +1382,6 @@ class Commande extends CommonOrder
 			$action = '';
 			$reshook = $hookmanager->executeHooks('createFrom', $parameters, $this, $action); // Note that $action and $object may have been modified by some hooks
 			if ($reshook < 0) {
-				$this->errors += $hookmanager->errors;
-				$this->error = $hookmanager->error;
 				$error++;
 			}
 
@@ -1827,7 +1796,7 @@ class Commande extends CommonOrder
 		$sql .= ", c.module_source, c.pos_source";
 		$sql .= ", i.libelle as label_incoterms";
 		$sql .= ', p.code as mode_reglement_code, p.libelle as mode_reglement_libelle';
-		$sql .= ', cr.code as cond_reglement_code, cr.libelle as cond_reglement_libelle, cr.libelle_facture as cond_reglement_libelle_doc';
+		$sql .= ', cr.code as cond_reglement_code, cr.libelle as cond_reglement_libelle, cr.libelle_facture as cond_reglement_libelle_doc, cr.deposit_percent as cond_reglement_deposit_percent';
 		$sql .= ', ca.code as availability_code, ca.label as availability_label';
 		$sql .= ', dr.code as demand_reason_code';
 		$sql .= ' FROM '.MAIN_DB_PREFIX.'commande as c';
@@ -1906,6 +1875,7 @@ class Commande extends CommonOrder
 				$this->cond_reglement_code	= $obj->cond_reglement_code;
 				$this->cond_reglement		= $obj->cond_reglement_libelle;
 				$this->cond_reglement_doc = $obj->cond_reglement_libelle_doc;
+				$this->cond_reglement_deposit_percent = $obj->cond_reglement_deposit_percent;
 				$this->fk_account = $obj->fk_account;
 				$this->availability_id = $obj->fk_availability;
 				$this->availability_code	= $obj->availability_code;
diff --git a/htdocs/commande/class/commandestats.class.php b/htdocs/commande/class/commandestats.class.php
index 8d55f9794ee..8ac89d5a18f 100644
--- a/htdocs/commande/class/commandestats.class.php
+++ b/htdocs/commande/class/commandestats.class.php
@@ -93,6 +93,9 @@ class CommandeStats extends Stats
 		//$this->where.= " AND c.fk_soc = s.rowid AND c.entity = ".$conf->entity;
 		$this->where .= ($this->where ? ' AND ' : '').'c.entity IN ('.getEntity('commande').')';
 
+		if (!$user->rights->societe->client->voir && !$this->socid) {
+			$this->where .= " AND c.fk_soc = sc.fk_soc AND sc.fk_user = ".((int) $user->id);
+		}
 		if ($this->socid) {
 			$this->where .= " AND c.fk_soc = ".((int) $this->socid);
 		}
@@ -126,7 +129,7 @@ class CommandeStats extends Stats
 		$sql = "SELECT date_format(c.date_commande,'%m') as dm, COUNT(*) as nb";
 		$sql .= " FROM ".$this->from;
 		if (!$user->rights->societe->client->voir && !$this->socid) {
-			$sql .= "  INNER JOIN ".MAIN_DB_PREFIX."societe_commerciaux as sc ON c.fk_soc = sc.fk_soc AND sc.fk_user = ".((int) $user->id);
+			$sql .= ", ".MAIN_DB_PREFIX."societe_commerciaux as sc";
 		}
 		$sql .= $this->join;
 		$sql .= " WHERE c.date_commande BETWEEN '".$this->db->idate(dol_get_first_day($year))."' AND '".$this->db->idate(dol_get_last_day($year))."'";
@@ -151,7 +154,7 @@ class CommandeStats extends Stats
 		$sql = "SELECT date_format(c.date_commande,'%Y') as dm, COUNT(*) as nb, SUM(c.".$this->field.")";
 		$sql .= " FROM ".$this->from;
 		if (!$user->rights->societe->client->voir && !$this->socid) {
-			$sql .= "  INNER JOIN ".MAIN_DB_PREFIX."societe_commerciaux as sc ON c.fk_soc = sc.fk_soc AND sc.fk_user = ".((int) $user->id);
+			$sql .= ", ".MAIN_DB_PREFIX."societe_commerciaux as sc";
 		}
 		$sql .= $this->join;
 		$sql .= " WHERE ".$this->where;
@@ -175,7 +178,7 @@ class CommandeStats extends Stats
 		$sql = "SELECT date_format(c.date_commande,'%m') as dm, SUM(c.".$this->field.")";
 		$sql .= " FROM ".$this->from;
 		if (!$user->rights->societe->client->voir && !$this->socid) {
-			$sql .= "  INNER JOIN ".MAIN_DB_PREFIX."societe_commerciaux as sc ON c.fk_soc = sc.fk_soc AND sc.fk_user = ".((int) $user->id);
+			$sql .= ", ".MAIN_DB_PREFIX."societe_commerciaux as sc";
 		}
 		$sql .= $this->join;
 		$sql .= " WHERE c.date_commande BETWEEN '".$this->db->idate(dol_get_first_day($year))."' AND '".$this->db->idate(dol_get_last_day($year))."'";
@@ -200,7 +203,7 @@ class CommandeStats extends Stats
 		$sql = "SELECT date_format(c.date_commande,'%m') as dm, AVG(c.".$this->field.")";
 		$sql .= " FROM ".$this->from;
 		if (!$user->rights->societe->client->voir && !$this->socid) {
-			$sql .= "  INNER JOIN ".MAIN_DB_PREFIX."societe_commerciaux as sc ON c.fk_soc = sc.fk_soc AND sc.fk_user = ".((int) $user->id);
+			$sql .= ", ".MAIN_DB_PREFIX."societe_commerciaux as sc";
 		}
 		$sql .= $this->join;
 		$sql .= " WHERE c.date_commande BETWEEN '".$this->db->idate(dol_get_first_day($year))."' AND '".$this->db->idate(dol_get_last_day($year))."'";
@@ -223,7 +226,7 @@ class CommandeStats extends Stats
 		$sql = "SELECT date_format(c.date_commande,'%Y') as year, COUNT(*) as nb, SUM(c.".$this->field.") as total, AVG(".$this->field.") as avg";
 		$sql .= " FROM ".$this->from;
 		if (!$user->rights->societe->client->voir && !$this->socid) {
-			$sql .= "  INNER JOIN ".MAIN_DB_PREFIX."societe_commerciaux as sc ON c.fk_soc = sc.fk_soc AND sc.fk_user = ".((int) $user->id);
+			$sql .= ", ".MAIN_DB_PREFIX."societe_commerciaux as sc";
 		}
 		$sql .= $this->join;
 		$sql .= " WHERE ".$this->where;
@@ -247,7 +250,7 @@ class CommandeStats extends Stats
 		$sql = "SELECT product.ref, COUNT(product.ref) as nb, SUM(tl.".$this->field_line.") as total, AVG(tl.".$this->field_line.") as avg";
 		$sql .= " FROM ".$this->from.", ".$this->from_line.", ".MAIN_DB_PREFIX."product as product";
 		if (!$user->rights->societe->client->voir && !$user->socid) {
-			$sql .= "  INNER JOIN ".MAIN_DB_PREFIX."societe_commerciaux as sc ON c.fk_soc = sc.fk_soc AND sc.fk_user = ".((int) $user->id);
+			$sql .= ", ".MAIN_DB_PREFIX."societe_commerciaux as sc";
 		}
 		$sql .= $this->join;
 		$sql .= " WHERE ".$this->where;
diff --git a/htdocs/commande/list.php b/htdocs/commande/list.php
index 07db0c6077c..09d0c7c7e48 100644
--- a/htdocs/commande/list.php
+++ b/htdocs/commande/list.php
@@ -45,6 +45,7 @@ require_once DOL_DOCUMENT_ROOT.'/commande/class/commande.class.php';
 require_once DOL_DOCUMENT_ROOT.'/compta/facture/class/facture.class.php';
 require_once DOL_DOCUMENT_ROOT.'/product/class/product.class.php';
 require_once DOL_DOCUMENT_ROOT.'/projet/class/project.class.php';
+require_once DOL_DOCUMENT_ROOT.'/fourn/class/fournisseur.commande.class.php';
 
 // Load translation files required by the page
 $langs->loadLangs(array("orders", 'sendings', 'deliveries', 'companies', 'compta', 'bills', 'stocks', 'products'));
@@ -70,6 +71,7 @@ $search_datedelivery_start = dol_mktime(0, 0, 0, GETPOST('search_datedelivery_st
 $search_datedelivery_end = dol_mktime(23, 59, 59, GETPOST('search_datedelivery_end_month', 'int'), GETPOST('search_datedelivery_end_day', 'int'), GETPOST('search_datedelivery_end_year', 'int'));
 $search_product_category = GETPOST('search_product_category', 'int');
 $search_ref = GETPOST('search_ref', 'alpha') != '' ?GETPOST('search_ref', 'alpha') : GETPOST('sref', 'alpha');
+$search_supplierorder = GETPOST('search_supplierorder', 'alpha');
 $search_ref_customer = GETPOST('search_ref_customer', 'alpha');
 $search_company = GETPOST('search_company', 'alpha');
 $search_company_alias = GETPOST('search_company_alias', 'alpha');
@@ -139,6 +141,7 @@ $show_shippable_command = GETPOST('show_shippable_command', 'aZ09');
 $object = new Commande($db);
 $hookmanager->initHooks(array('orderlist'));
 $extrafields = new ExtraFields($db);
+$supplierorder = new CommandeFournisseur($db);
 
 // fetch optionals attributes and labels
 $extrafields->fetch_name_optionals_label($object->table_element);
@@ -147,6 +150,7 @@ $search_array_options = $extrafields->getOptionalsFromPost($object->table_elemen
 // List of fields to search into when doing a "search in all"
 $fieldstosearchall = array(
 	'c.ref'=>'Ref',
+	'cf.ref'=>'SupplierOrder',
 	'c.ref_client'=>'RefCustomerOrder',
 	'pd.description'=>'Description',
 	's.nom'=>"ThirdParty",
@@ -161,7 +165,9 @@ if (empty($user->socid)) {
 
 $checkedtypetiers = 0;
 $arrayfields = array(
+	'ec.fk_socpeople'=>array('label'=>"BySalesRepresentative", 'checked'=>1, 'position'=>1),
 	'c.ref'=>array('label'=>"Ref", 'checked'=>1, 'position'=>5),
+	'cf.ref'=>array('label'=>"SupplierOrder", 'checked'=>1, 'position'=>6),
 	'c.ref_client'=>array('label'=>"RefCustomerOrder", 'checked'=>-1, 'position'=>10),
 	'p.ref'=>array('label'=>"ProjectRef", 'checked'=>-1, 'enabled'=>(empty($conf->projet->enabled) ? 0 : 1), 'position'=>20),
 	'p.title'=>array('label'=>"ProjectLabel", 'checked'=>0, 'enabled'=>(empty($conf->projet->enabled) ? 0 : 1), 'position'=>25),
@@ -232,6 +238,7 @@ if (empty($reshook)) {
 		$search_sale = '';
 		$search_product_category = '';
 		$search_ref = '';
+		$search_supplierorder = "";
 		$search_ref_customer = '';
 		$search_company = '';
 		$search_company_alias = '';
@@ -337,15 +344,15 @@ if ($action == 'shipped' && $permissiontoadd) {
 		$error = 0;
 		foreach ($toselect as $checked) {
 			if ($objecttmp->fetch($checked)) {
-				if ($objecttmp->statut == 1 || $objecttmp->statut == 2) {
+				if ($objecttmp->statut == 1) {
 					if ($objecttmp->cloture($user)) {
-						setEventMessage($objecttmp->ref." ".$langs->trans('PassedInShippedStatus'), 'mesgs');
+						setEventMessage($objecttmp->ref." ".$langs->trans('PassedInOpenStatus'), 'mesgs');
 					} else {
-						setEventMessage($langs->trans('YouCantShipThis'), 'errors');
+						setEventMessage($langs->trans('CantBeValidated'), 'errors');
 						$error++;
 					}
 				} else {
-					setEventMessage($objecttmp->ref." ".$langs->trans('MustBeValidatedBefore'), 'errors');
+					setEventMessage($objecttmp->ref." ".$langs->trans('IsNotADraft'), 'errors');
 					$error++;
 				}
 			} else {
@@ -422,10 +429,12 @@ $sql .= " typent.code as typent_code,";
 $sql .= " state.code_departement as state_code, state.nom as state_name,";
 $sql .= " country.code as country_code,";
 $sql .= ' c.rowid, c.ref, c.total_ht, c.total_tva, c.total_ttc, c.ref_client, c.fk_user_author,';
+$sql .= " cf.rowid as orderid, cf.ref as supplierorder,";
 $sql .= ' c.fk_multicurrency, c.multicurrency_code, c.multicurrency_tx, c.multicurrency_total_ht, c.multicurrency_total_tva as multicurrency_total_vat, c.multicurrency_total_ttc,';
 $sql .= ' c.date_valid, c.date_commande, c.note_public, c.note_private, c.date_livraison as date_delivery, c.fk_statut, c.facture as billed,';
 $sql .= ' c.date_creation as date_creation, c.tms as date_update, c.date_cloture as date_cloture,';
 $sql .= ' p.rowid as project_id, p.ref as project_ref, p.title as project_label,';
+$sql .= ' ec.fk_socpeople,';
 $sql .= ' u.login, u.lastname, u.firstname, u.email as user_email, u.statut as user_statut, u.entity, u.photo, u.office_phone, u.office_fax, u.user_mobile, u.job, u.gender,';
 $sql .= ' c.fk_cond_reglement,c.fk_mode_reglement,c.fk_shipping_method,';
 $sql .= ' c.fk_input_reason';
@@ -462,15 +471,24 @@ if ($search_product_category > 0) {
 $sql .= " LEFT JOIN ".MAIN_DB_PREFIX."projet as p ON p.rowid = c.fk_projet";
 $sql .= ' LEFT JOIN '.MAIN_DB_PREFIX.'user as u ON c.fk_user_author = u.rowid';
 
+$sql .= ' LEFT JOIN (';
+$sql .= " SELECT ee1.fk_source as fk_source, ee1.fk_target as fk_target FROM ".MAIN_DB_PREFIX."element_element as ee1 WHERE (ee1.sourcetype = 'order_supplier' AND ee1.targettype = 'commande')";
+$sql .= ' UNION ';
+$sql .= " SELECT ee2.fk_target as fk_source, ee2.fk_source as fk_target FROM ".MAIN_DB_PREFIX."element_element as ee2 WHERE (ee2.targettype = 'order_supplier' AND ee2.sourcetype = 'commande')";
+$sql .= ' ) as eecommande ON (c.rowid = eecommande.fk_target)';
+$sql .= " LEFT JOIN ".MAIN_DB_PREFIX."commande_fournisseur as cf on (cf.rowid = eecommande.fk_source)";
+
 // We'll need this table joined to the select in order to filter by sale
 if ($search_sale > 0 || (!$user->rights->societe->client->voir && !$socid)) {
 	$sql .= ", ".MAIN_DB_PREFIX."societe_commerciaux as sc";
 }
-if ($search_user > 0) {
-	$sql .= ", ".MAIN_DB_PREFIX."element_contact as ec";
-	$sql .= ", ".MAIN_DB_PREFIX."c_type_contact as tc";
-}
-
+// if ($search_user > 0) {	//deprecated to show sales representative
+$sql .= " LEFT JOIN (";
+$sql .= " SELECT ec1.element_id, ec1.fk_socpeople FROM ".MAIN_DB_PREFIX."element_contact as ec1";
+$sql .= " LEFT JOIN ".MAIN_DB_PREFIX."c_type_contact as tc ON ec1.fk_c_type_contact = tc.rowid";
+$sql .= " WHERE tc.element='commande' AND tc.source='internal'";
+$sql .= " ) AS ec on ec.element_id = c.rowid";
+// } 	//deprecated to show sales representative
 // Add table from hooks
 $parameters = array();
 $reshook = $hookmanager->executeHooks('printFieldListFrom', $parameters, $object); // Note that $action and $object may have been modified by hook
@@ -490,6 +508,9 @@ if (!$user->rights->societe->client->voir && !$socid) {
 if ($search_ref) {
 	$sql .= natural_search('c.ref', $search_ref);
 }
+if ($search_supplierorder) {
+	$sql .= natural_search('cf.ref', $search_refsupplier);
+}
 if ($search_ref_customer) {
 	$sql .= natural_search('c.ref_client', $search_ref_customer);
 }
@@ -564,7 +585,7 @@ if ($search_sale > 0) {
 	$sql .= " AND s.rowid = sc.fk_soc AND sc.fk_user = ".((int) $search_sale);
 }
 if ($search_user > 0) {
-	$sql .= " AND ec.fk_c_type_contact = tc.rowid AND tc.element='commande' AND tc.source='internal' AND ec.element_id = c.rowid AND ec.fk_socpeople = ".((int) $search_user);
+	$sql .= " AND ec.fk_socpeople = ".((int) $search_user);
 }
 if ($search_total_ht != '') {
 	$sql .= natural_search('c.total_ht', $search_total_ht, 1);
@@ -738,6 +759,9 @@ if ($resql) {
 	if ($search_ref) {
 		$param .= '&search_ref='.urlencode($search_ref);
 	}
+	if ($search_supplierorder) {
+		$param .= '&search_supplierorder='.urlencode($search_supplierorder);
+	}
 	if ($search_company) {
 		$param .= '&search_company='.urlencode($search_company);
 	}
@@ -1031,12 +1055,24 @@ if ($resql) {
 	print '<table class="tagtable liste'.($moreforfilter ? " listwithfilterbefore" : "").'">'."\n";
 
 	print '<tr class="liste_titre_filter">';
+	// Salesperson
+	if (!empty($arrayfields['ec.fk_socpeople']['checked'])) {
+		print '<td class="liste_titre">';
+		print '&nbsp;';
+		print '</td>';
+	}
 	// Ref
 	if (!empty($arrayfields['c.ref']['checked'])) {
 		print '<td class="liste_titre">';
 		print '<input class="flat" size="6" type="text" name="search_ref" value="'.dol_escape_htmltag($search_ref).'">';
 		print '</td>';
 	}
+	// Supplier Order
+	if (!empty($arrayfields['cf.ref']['checked'])) {
+		print '<td class="liste_titre">';
+		print '<input class="flat maxwidth50" type="text" name="search_supplierorder" value="'.$search_supplierorder.'">';
+		print '</td>';
+	}
 	// Ref customer
 	if (!empty($arrayfields['c.ref_client']['checked'])) {
 		print '<td class="liste_titre" align="left">';
@@ -1267,9 +1303,15 @@ if ($resql) {
 
 	// Fields title
 	print '<tr class="liste_titre">';
+	if (!empty($arrayfields['ec.fk_socpeople']['checked'])) {
+		print_liste_field_titre($arrayfields['ec.fk_socpeople']['label'], $_SERVER["PHP_SELF"], 'ec.fk_socpeople', '', $param, '', $sortfield, $sortorder);
+	}
 	if (!empty($arrayfields['c.ref']['checked'])) {
 		print_liste_field_titre($arrayfields['c.ref']['label'], $_SERVER["PHP_SELF"], 'c.ref', '', $param, '', $sortfield, $sortorder);
 	}
+	if (!empty($arrayfields['cf.ref']['checked'])) {
+		print_liste_field_titre($arrayfields['cf.ref']['label'], $_SERVER["PHP_SELF"], 'cf.ref', '', $param, '', $sortfield, $sortorder);
+	}
 	if (!empty($arrayfields['c.ref_client']['checked'])) {
 		print_liste_field_titre($arrayfields['c.ref_client']['label'], $_SERVER["PHP_SELF"], 'c.ref_client', '', $param, '', $sortfield, $sortorder);
 	}
@@ -1388,6 +1430,7 @@ if ($resql) {
 	$generic_commande = new Commande($db);
 	$generic_product = new Product($db);
 	$userstatic = new User($db);
+	$salespersonstatic = new User($db);
 	$i = 0;
 	$totalarray = array('nbfield' => 0, 'val' => array(), 'pos' => array());
 	while ($i < min($num, $limit)) {
@@ -1414,6 +1457,9 @@ if ($resql) {
 			$getNomUrl_cache[$obj->socid] = $companystatic->getNomUrl(1, 'customer');
 		}
 
+		$supplierorder->id = $obj->orderid;
+		$supplierorder->ref = $obj->supplierorder;
+
 		$generic_commande->id = $obj->rowid;
 		$generic_commande->ref = $obj->ref;
 		$generic_commande->statut = $obj->fk_statut;
@@ -1425,6 +1471,9 @@ if ($resql) {
 		$generic_commande->total_ht = $obj->total_ht;
 		$generic_commande->total_tva = $obj->total_tva;
 		$generic_commande->total_ttc = $obj->total_ttc;
+		$generic_commande->multicurrency_total_ht = $obj->multicurrency_total_ht;
+		$generic_commande->multicurrency_total_tva = $obj->multicurrency_total_tva;
+		$generic_commande->multicurrency_total_ttc = $obj->multicurrency_total_ttc;
 		$generic_commande->note_public = $obj->note_public;
 		$generic_commande->note_private = $obj->note_private;
 
@@ -1433,7 +1482,21 @@ if ($resql) {
 		$projectstatic->title = $obj->project_label;
 
 		print '<tr class="oddeven">';
-
+		// Salesperson
+		$salespersonstatic->id = $obj->fk_socpeople;
+		$salespersonstatic->fetch($salespersonstatic->id);
+		if (!empty($arrayfields['ec.fk_socpeople']['checked'])) {
+			print '<td class="tdoverflowmax150">';
+			if ($salespersonstatic->id) {
+				print $salespersonstatic->getNomUrl(-1);
+			} else {
+				print '&nbsp;';
+			}
+			print "</td>\n";
+			if (!$i) {
+				$totalarray['nbfield']++;
+			}
+		}
 		// Ref
 		if (!empty($arrayfields['c.ref']['checked'])) {
 			print '<td class="nowraponall">';
@@ -1455,9 +1518,19 @@ if ($resql) {
 			}
 		}
 
+		// Ref Order
+		if (!empty($arrayfields['cf.ref']['checked'])) {
+			print "<td>";
+			print $supplierorder->id ? $supplierorder->getNomUrl(1) : '';
+			print "</td>\n";
+			if (!$i) {
+				$totalarray['nbfield']++;
+			}
+		}
+			
 		// Ref customer
 		if (!empty($arrayfields['c.ref_client']['checked'])) {
-			print '<td class="nowrap tdoverflowmax200">'.$obj->ref_client.'</td>';
+			print '<td class="nowrap tdoverflowmax150">'.$obj->ref_client.'</td>';
 			if (!$i) {
 				$totalarray['nbfield']++;
 			}
@@ -1678,12 +1751,24 @@ if ($resql) {
 			if (!$i) {
 				$totalarray['nbfield']++;
 			}
+			if (!$i) {
+				$totalarray['pos'][$totalarray['nbfield']] = 'c.multicurrency_total_ht';
+			}
+			if (isset($totalarray['val']['c.multicurrency_total_ht'])) {
+				$totalarray['val']['c.multicurrency_total_ht'] += $obj->multicurrency_total_ht;
+			} else {
+				$totalarray['val']['c.multicurrency_total_ht'] = $obj->multicurrency_total_ht;
+			}
 		}
 		// Amount VAT
 		if (!empty($arrayfields['c.multicurrency_total_vat']['checked'])) {
 			print '<td class="right nowrap"><span class="amount">'.price($obj->multicurrency_total_vat)."</span></td>\n";
 			if (!$i) {
 				$totalarray['nbfield']++;
+				if (!$i) {
+					$totalarray['pos'][$totalarray['nbfield']] = 'c.multicurrency_total_tva';
+				}
+				$totalarray['val']['c.multicurrency_total_tva'] += $obj->multicurrency_total_tva;
 			}
 		}
 		// Amount TTC
@@ -1692,6 +1777,10 @@ if ($resql) {
 			if (!$i) {
 				$totalarray['nbfield']++;
 			}
+			if (!$i) {
+				$totalarray['pos'][$totalarray['nbfield']] = 'c.multicurrency_total_ttc';
+			}
+			$totalarray['val']['c.multicurrency_total_ttc'] += $obj->multicurrency_total_ttc;
 		}
 
 		$userstatic->id = $obj->fk_user_author;
@@ -1710,7 +1799,7 @@ if ($resql) {
 
 		// Author
 		if (!empty($arrayfields['u.login']['checked'])) {
-			print '<td class="tdoverflowmax200">';
+			print '<td class="tdoverflowmax150">';
 			if ($userstatic->id) {
 				print $userstatic->getNomUrl(-1);
 			} else {
@@ -1762,7 +1851,7 @@ if ($resql) {
 		// Note public
 		if (!empty($arrayfields['c.note_public']['checked'])) {
 			print '<td align="center" class="nowrap">';
-			print dol_string_nohtmltag($obj->note_public);
+			print dol_escape_htmltag($obj->note_public);
 			print '</td>';
 			if (!$i) {
 				$totalarray['nbfield']++;
@@ -1772,7 +1861,7 @@ if ($resql) {
 		// Note private
 		if (!empty($arrayfields['c.note_private']['checked'])) {
 			print '<td align="center" class="nowrap">';
-			print dol_string_nohtmltag($obj->note_private);
+			print dol_escape_htmltag($obj->note_private);
 			print '</td>';
 			if (!$i) {
 				$totalarray['nbfield']++;
@@ -1911,8 +2000,6 @@ if ($resql) {
 
 		print "</tr>\n";
 
-		$total += $obj->total_ht;
-		$subtotal += $obj->total_ht;
 		$i++;
 	}
 
diff --git a/htdocs/commande/stats/index.php b/htdocs/commande/stats/index.php
index 7d0f9fb5a3a..bd92dc084c7 100644
--- a/htdocs/commande/stats/index.php
+++ b/htdocs/commande/stats/index.php
@@ -45,14 +45,8 @@ if ($mode == 'customer' && !$user->rights->commande->lire) {
 if ($mode == 'supplier' && !$user->rights->fournisseur->commande->lire) {
 	accessforbidden();
 }
-if ($mode == 'supplier') {
-	$object_status = GETPOST('object_status', 'array:int');
-	$object_status = implode(',', $object_status);
-} else {
-	$object_status = GETPOST('object_status', 'intcomma');
-}
-
 
+$object_status = GETPOST('object_status', 'intcomma');
 $typent_id = GETPOST('typent_id', 'int');
 $categ_id = GETPOST('categ_id', 'categ_id');
 
diff --git a/htdocs/commande/tpl/linkedobjectblock.tpl.php b/htdocs/commande/tpl/linkedobjectblock.tpl.php
index a0641c4c493..eeb7e5d5ad2 100644
--- a/htdocs/commande/tpl/linkedobjectblock.tpl.php
+++ b/htdocs/commande/tpl/linkedobjectblock.tpl.php
@@ -37,6 +37,9 @@ $langs->load("orders");
 $linkedObjectBlock = dol_sort_array($linkedObjectBlock, 'date', 'desc', 0, 0, 1);
 
 $total = 0;
+if (!empty($conf->multicurrency->enabled)) {
+    $multicurrency_total = 0;
+}
 $ilink = 0;
 foreach ($linkedObjectBlock as $key => $objectlink) {
 	$ilink++;
@@ -56,8 +59,12 @@ foreach ($linkedObjectBlock as $key => $objectlink) {
 	echo '<td class="linkedcol-date" align="center">'.dol_print_date($objectlink->date, 'day').'</td>';
 	echo '<td class="linkedcol-amount right">';
 	if ($user->rights->commande->lire) {
+		if (!empty($conf->multicurrency->enabled) && !empty($objectlink->multicurrency_code) && ($conf->currency!=$objectlink->multicurrency_code)) {
+			$multicurrency_total = $multicurrency_total + $objectlink->multicurrency_total_ht;
+            echo $objectlink->multicurrency_code.' '.price($objectlink->multicurrency_total_ht).'<br>';
+		}
 		$total = $total + $objectlink->total_ht;
-		echo price($objectlink->total_ht);
+		echo $conf->currency.' '.price($objectlink->total_ht);
 	}
 	echo '</td>';
 	echo '<td class="linkedcol-statut right">'.$objectlink->getLibStatut(3).'</td>';
@@ -75,7 +82,7 @@ if (count($linkedObjectBlock) > 1) {
 	echo '<td></td>';
 	echo '<td class="center"></td>';
 	echo '<td class="center"></td>';
-	echo '<td class="right">'.price($total).'</td>';
+	echo '<td class="right">'.$conf->currency.' '.price($total).'</td>';
 	echo '<td class="right"></td>';
 	echo '<td class="right"></td>';
 	echo "</tr>\n";
