diff --git a/htdocs/admin/debugbar.php b/htdocs/admin/debugbar.php
index 6f77e1ef4e9..111106a9c69 100644
--- a/htdocs/admin/debugbar.php
+++ b/htdocs/admin/debugbar.php
@@ -101,18 +101,14 @@ print '<td>'.$langs->trans("Parameter").'</td><td></td>';
 print '<td class="right"><input type="submit" class="button button-edit" value="'.$langs->trans("Modify").'"></td>';
 print "</tr>\n";
 
+print '<tr class="oddeven"><td class="nowrap">'.$langs->trans("DEBUGBAR_LOGS_LINES_NUMBER").'</td>';
+print '<td colspan="2"><input type="text" class="flat width75" name="DEBUGBAR_LOGS_LINES_NUMBER" value="'.(!getDolGlobalString('DEBUGBAR_LOGS_LINES_NUMBER') ? 250 : $conf->global->DEBUGBAR_LOGS_LINES_NUMBER).'">'; // This slow seriously output
+print ' <span class="opacitymedium">'.$langs->trans("WarningValueHigherSlowsDramaticalyOutput").'</span></td></tr>';
+
 print '<tr class="oddeven"><td class="nowrap">'.$langs->trans("DEBUGBAR_USE_LOG_FILE").'</td>';
-print '<td>';
+print '<td colspan="2">';
 print $form->selectyesno('DEBUGBAR_USE_LOG_FILE', getDolGlobalString('DEBUGBAR_USE_LOG_FILE'), 1);
-print '</td><td>';
-print '<span class="opacitymedium"> '.$langs->trans("UsingLogFileShowAllRecordOfSubrequestButIsSlower").'</span>';
-print '</td></tr>';
-
-print '<tr class="oddeven"><td class="nowrap">'.$langs->trans("DEBUGBAR_LOGS_LINES_NUMBER").'</td>';
-print '<td><input type="text" class="flat width75" name="DEBUGBAR_LOGS_LINES_NUMBER" value="'.(!getDolGlobalString('DEBUGBAR_LOGS_LINES_NUMBER') ? 250 : $conf->global->DEBUGBAR_LOGS_LINES_NUMBER).'">'; // This slow seriously output
-print '</td><td>';
-print '<span class="opacitymedium">'.$langs->trans("WarningValueHigherSlowsDramaticalyOutput").'</span>';
-print '</td></tr>';
+print ' <span class="opacitymedium">'.$langs->trans("UsingLogFileShowAllRecordOfSubrequestButIsSlower").'</span></td></tr>';
 
 print '</table>';
 print "</form>\n";
diff --git a/htdocs/admin/dict.php b/htdocs/admin/dict.php
index fa68a87391a..cf6b7584173 100644
--- a/htdocs/admin/dict.php
+++ b/htdocs/admin/dict.php
@@ -1102,7 +1102,7 @@ if (empty($reshook)) {
 			//print $sql;
 			$resql = $db->query($sql);
 			if (!$resql) {
-				setEventMessages($db->error(), null, 'errors');
+				dol_print_error($db);
 			}
 		}
 
diff --git a/htdocs/admin/mails_templates.php b/htdocs/admin/mails_templates.php
index c51b668cae4..024b9b9e28b 100644
--- a/htdocs/admin/mails_templates.php
+++ b/htdocs/admin/mails_templates.php
@@ -221,7 +221,7 @@ if (isModEnabled("shipping")) {
 	$elementList['shipping_send'] = img_picto('', 'dolly', 'class="pictofixedwidth"').dol_escape_htmltag($langs->trans('MailToSendShipment'));
 }
 if (isModEnabled("reception")) {
-	$elementList['reception_send'] = img_picto('', 'dollyrevert', 'class="pictofixedwidth"').dol_escape_htmltag($langs->trans('MailToSendReception'));
+	$elementList['reception_send'] = img_picto('', 'dolly', 'class="pictofixedwidth"').dol_escape_htmltag($langs->trans('MailToSendReception'));
 }
 if (isModEnabled('intervention')) {
 	$elementList['fichinter_send'] = img_picto('', 'intervention', 'class="pictofixedwidth"').dol_escape_htmltag($langs->trans('MailToSendIntervention'));
