diff --git a/htdocs/comm/propal/tpl/linkedobjectblock.tpl.php b/htdocs/comm/propal/tpl/linkedobjectblock.tpl.php
index b1a6ac11bae..30de1ea7f6a 100644
--- a/htdocs/comm/propal/tpl/linkedobjectblock.tpl.php
+++ b/htdocs/comm/propal/tpl/linkedobjectblock.tpl.php
@@ -46,6 +46,9 @@ $linkedObjectBlock = dol_sort_array($linkedObjectBlock, 'date', 'desc', 0, 0, 1)
 '@phan-var-force CommonObject[] $linkedObjectBlock';  // Repeat because type lost after dol_sort_array)
 
 $total = 0;
+if (isModEnabled('multicurrency')) {
+	$multicurrency_total = 0;
+}
 $ilink = 0;
 foreach ($linkedObjectBlock as $key => $objectlink) {
 	$ilink++;
@@ -66,8 +69,12 @@ foreach ($linkedObjectBlock as $key => $objectlink) {
 	print '<td class="linkedcol-date center">'.dol_print_date($objectlink->date, 'day').'</td>';
 	print '<td class="linkedcol-amount right">';
 	if ($user->hasRight('propal', 'lire')) {
+		if (isModEnabled('multicurrency') && !empty($objectlink->multicurrency_code) && $conf->currency != $objectlink->multicurrency_code) {
+			$multicurrency_total += $objectlink->multicurrency_total_ht;
+			echo $objectlink->multicurrency_code.' '.price($objectlink->multicurrency_total_ht).'<br>';
+		}
 		$total += $objectlink->total_ht;
-		echo price($objectlink->total_ht);
+		echo $conf->currency.' '.price($objectlink->total_ht);
 	}
 	print '</td>';
 	print '<td class="linkedcol-statut right">'.$objectlink->getLibStatut(3).'</td>';
@@ -80,7 +87,7 @@ if (count($linkedObjectBlock) > 1) {
 	print '<td></td>';
 	print '<td class="center"></td>';
 	print '<td class="center"></td>';
-	print '<td class="right">'.price($total).'</td>';
+	print '<td class="right">'.$conf->currency.' '.price($total).'</td>';
 	print '<td class="right"></td>';
 	print '<td class="right"></td>';
 	print "</tr>\n";
